<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 3 Analysing Data | The pavo handbook: studying biological colouration in R</title>
  <meta name="description" content="Chapter 3 Analysing Data | The pavo handbook: studying biological colouration in R" />
  <meta name="generator" content="bookdown 0.21.6 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 3 Analysing Data | The pavo handbook: studying biological colouration in R" />
  <meta property="og:type" content="book" />
  
  
  
  <meta name="github-repo" content="bisaloo/colsci-book" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 Analysing Data | The pavo handbook: studying biological colouration in R" />
  
  
  



<meta name="date" content="2021-02-13" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="importing-processing-and-visualising-data.html"/>
<link rel="next" href="spectral-shape-descriptors.html"/>
<script src="libs/header-attrs-2.6.6/header-attrs.js"></script>
<script src="libs/jquery-3.5.1/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<script src="libs/htmlwidgets-1.5.3/htmlwidgets.js"></script>
<script src="libs/rglWebGL-binding-0.104.16/rglWebGL.js"></script>
<link href="libs/rglwidgetClass-0.104.16/rgl.css" rel="stylesheet" />
<script src="libs/rglwidgetClass-0.104.16/rglClass.min.js"></script>
<script src="libs/CanvasMatrix4-0.104.16/CanvasMatrix.min.js"></script>
<link href="libs/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet" />
<script src="libs/crosstalk-1.1.1/js/crosstalk.min.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">The pavo handbook: studying biological colouration in R</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Package Overview</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#classes-and-attributes"><i class="fa fa-check"></i><b>1.1</b> Classes and Attributes</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#suggestions-and-assistance"><i class="fa fa-check"></i><b>1.2</b> Suggestions and Assistance</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#citation-of-methods-implemented-in-pavo"><i class="fa fa-check"></i><b>1.3</b> Citation of methods implemented in <code>pavo</code></a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#acknowledgements"><i class="fa fa-check"></i><b>1.4</b> Acknowledgements</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="importing-processing-and-visualising-data.html"><a href="importing-processing-and-visualising-data.html"><i class="fa fa-check"></i><b>2</b> Importing, Processing, and Visualising Data</a>
<ul>
<li class="chapter" data-level="2.1" data-path="importing-processing-and-visualising-data.html"><a href="importing-processing-and-visualising-data.html#organizing-spectral-data"><i class="fa fa-check"></i><b>2.1</b> Organizing Spectral Data</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="importing-processing-and-visualising-data.html"><a href="importing-processing-and-visualising-data.html#spectral-dataset-description"><i class="fa fa-check"></i><b>2.1.1</b> Spectral Dataset Description</a></li>
<li class="chapter" data-level="2.1.2" data-path="importing-processing-and-visualising-data.html"><a href="importing-processing-and-visualising-data.html#importing"><i class="fa fa-check"></i><b>2.1.2</b> Importing</a></li>
<li class="chapter" data-level="2.1.3" data-path="importing-processing-and-visualising-data.html"><a href="importing-processing-and-visualising-data.html#subsetting-and-merging-spectral-data"><i class="fa fa-check"></i><b>2.1.3</b> Subsetting and Merging Spectral Data</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="importing-processing-and-visualising-data.html"><a href="importing-processing-and-visualising-data.html#processing-spectral-data"><i class="fa fa-check"></i><b>2.2</b> Processing Spectral Data</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="importing-processing-and-visualising-data.html"><a href="importing-processing-and-visualising-data.html#averaging-spectra"><i class="fa fa-check"></i><b>2.2.1</b> Averaging Spectra</a></li>
<li class="chapter" data-level="2.2.2" data-path="importing-processing-and-visualising-data.html"><a href="importing-processing-and-visualising-data.html#normalizing-and-smoothing-spectra"><i class="fa fa-check"></i><b>2.2.2</b> Normalizing and Smoothing Spectra</a></li>
<li class="chapter" data-level="2.2.3" data-path="importing-processing-and-visualising-data.html"><a href="importing-processing-and-visualising-data.html#binning-and-pca-analysis-of-spectral-shape"><i class="fa fa-check"></i><b>2.2.3</b> Binning and PCA Analysis of Spectral Shape</a></li>
<li class="chapter" data-level="2.2.4" data-path="importing-processing-and-visualising-data.html"><a href="importing-processing-and-visualising-data.html#dealing-with-negative-values-in-spectra"><i class="fa fa-check"></i><b>2.2.4</b> Dealing With Negative Values in Spectra</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="importing-processing-and-visualising-data.html"><a href="importing-processing-and-visualising-data.html#visualizing-spectral-data"><i class="fa fa-check"></i><b>2.3</b> Visualizing Spectral Data</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="importing-processing-and-visualising-data.html"><a href="importing-processing-and-visualising-data.html#the-plot-function-options"><i class="fa fa-check"></i><b>2.3.1</b> The <code>plot</code> Function Options</a></li>
<li class="chapter" data-level="2.3.2" data-path="importing-processing-and-visualising-data.html"><a href="importing-processing-and-visualising-data.html#the-aggplot-function"><i class="fa fa-check"></i><b>2.3.2</b> The <code>aggplot()</code> Function</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="importing-processing-and-visualising-data.html"><a href="importing-processing-and-visualising-data.html#organizing-spatial-image-data"><i class="fa fa-check"></i><b>2.4</b> Organizing Spatial (image) Data</a></li>
<li class="chapter" data-level="2.5" data-path="importing-processing-and-visualising-data.html"><a href="importing-processing-and-visualising-data.html#visualising-image-data"><i class="fa fa-check"></i><b>2.5</b> Visualising Image Data</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="importing-processing-and-visualising-data.html"><a href="importing-processing-and-visualising-data.html#the-plot-and-summary-functions"><i class="fa fa-check"></i><b>2.5.1</b> The <code>plot</code> and <code>summary</code> Functions</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="importing-processing-and-visualising-data.html"><a href="importing-processing-and-visualising-data.html#processing-image-data"><i class="fa fa-check"></i><b>2.6</b> Processing Image Data</a>
<ul>
<li class="chapter" data-level="2.6.1" data-path="importing-processing-and-visualising-data.html"><a href="importing-processing-and-visualising-data.html#overview"><i class="fa fa-check"></i><b>2.6.1</b> Overview</a></li>
<li class="chapter" data-level="2.6.2" data-path="importing-processing-and-visualising-data.html"><a href="importing-processing-and-visualising-data.html#setting-scales-and-defining-objects"><i class="fa fa-check"></i><b>2.6.2</b> Setting Scales and Defining Objects</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="analysing-data.html"><a href="analysing-data.html"><i class="fa fa-check"></i><b>3</b> Analysing Data</a>
<ul>
<li class="chapter" data-level="3.1" data-path="analysing-data.html"><a href="analysing-data.html#analysing-spectral-data"><i class="fa fa-check"></i><b>3.1</b> Analysing Spectral Data</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="analysing-data.html"><a href="analysing-data.html#spectral-dataset-description-1"><i class="fa fa-check"></i><b>3.1.1</b> Spectral Dataset Description</a></li>
<li class="chapter" data-level="3.1.2" data-path="analysing-data.html"><a href="analysing-data.html#overview-1"><i class="fa fa-check"></i><b>3.1.2</b> Overview</a></li>
<li class="chapter" data-level="3.1.3" data-path="analysing-data.html"><a href="analysing-data.html#spectral-shape-analysis"><i class="fa fa-check"></i><b>3.1.3</b> Spectral Shape Analysis</a></li>
<li class="chapter" data-level="3.1.4" data-path="analysing-data.html"><a href="analysing-data.html#visual-modelling"><i class="fa fa-check"></i><b>3.1.4</b> Visual Modelling</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="analysing-data.html"><a href="analysing-data.html#analysing-spatial-data"><i class="fa fa-check"></i><b>3.2</b> Analysing Spatial Data</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="analysing-data.html"><a href="analysing-data.html#image-based-colour-classification"><i class="fa fa-check"></i><b>3.2.1</b> Image-based colour classification</a></li>
<li class="chapter" data-level="3.2.2" data-path="analysing-data.html"><a href="analysing-data.html#adjacency-and-boundary-strength-analyses-and-overall-pattern-contrasts"><i class="fa fa-check"></i><b>3.2.2</b> Adjacency and Boundary Strength Analyses, and Overall Pattern Contrasts</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="spectral-shape-descriptors.html"><a href="spectral-shape-descriptors.html"><i class="fa fa-check"></i><b>4</b> Spectral shape descriptors</a></li>
<li class="chapter" data-level="5" data-path="colour-volume-estimation.html"><a href="colour-volume-estimation.html"><i class="fa fa-check"></i><b>5</b> Colour volume estimation</a>
<ul>
<li class="chapter" data-level="5.1" data-path="colour-volume-estimation.html"><a href="colour-volume-estimation.html#alpha-shapes-in-chromaticity-diagrams"><i class="fa fa-check"></i><b>5.1</b> <span class="math inline">\(\alpha\)</span>-shapes in chromaticity diagrams</a></li>
<li class="chapter" data-level="5.2" data-path="colour-volume-estimation.html"><a href="colour-volume-estimation.html#alpha-shapes-in-perceptually-uniform-spaces"><i class="fa fa-check"></i><b>5.2</b> <span class="math inline">\(\alpha\)</span>-shapes in perceptually uniform spaces</a></li>
<li class="chapter" data-level="5.3" data-path="colour-volume-estimation.html"><a href="colour-volume-estimation.html#colour-volume-overlap-with-alpha-shapes"><i class="fa fa-check"></i><b>5.3</b> Colour volume overlap with <span class="math inline">\(\alpha\)</span>-shapes</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">The pavo handbook: studying biological colouration in R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="analysing-data" class="section level1" number="3">
<h1><span class="header-section-number">Chapter 3</span> Analysing Data</h1>
<p>Let’s again begin by loading the package.</p>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb98-1"><a href="analysing-data.html#cb98-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the package, and set a global random-number seed for the reproducible generation of fake data later on.</span></span>
<span id="cb98-2"><a href="analysing-data.html#cb98-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pavo)</span>
<span id="cb98-3"><a href="analysing-data.html#cb98-3" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1612217</span>)</span></code></pre></div>
<div id="analysing-spectral-data" class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> Analysing Spectral Data</h2>
<div id="spectral-dataset-description-1" class="section level3" number="3.1.1">
<h3><span class="header-section-number">3.1.1</span> Spectral Dataset Description</h3>
<p>The raw spectral data used in this example are available from the package repository on <a href="https://github.com/rmaia/pavo">github</a>, located <a href="https://github.com/rmaia/pavo/tree/master/data_external/vignette">here</a>. You can download and extract it to follow the vignette exactly. Alternatively, the data are included as an RData file as part of the package installation, and so can be loaded directly (see below).</p>
<p>The data consist of reflectance spectra, obtained using Avantes equipment and software, from seven bird species: Northern Cardinal <em>Cardinalis cardinalis</em>, Wattled Jacana <em>Jacana jacana</em>, Baltimore Oriole <em>Icterus galbula</em>, Peach-fronted Parakeet <em>Aratinga aurea</em>, American Robin <em>Turdus migratorius</em>, and Sayaca Tanager <em>Thraupis sayaca</em>. Several individuals were measured (sample size varies by species), and 3 spectra were collected from each individual. However, the number of individuals measured per species is uneven and the data have additional peculiarities that should emphasize the flexibility <code>pavo</code> offers, as we’ll see below.</p>
<p>In addition, <code>pavo</code> includes three datasets that can be called with the <code>data</code> function. <code>data(teal)</code>, <code>data(sicalis)</code>, and <code>data(flowers)</code> will all be used in this vignette. See the help files for each dataset for more information; via <code>?teal</code>, <code>?sicalis</code>, and <code>?flowers</code>.</p>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb99-1"><a href="analysing-data.html#cb99-1" aria-hidden="true" tabindex="-1"></a>specs <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata/specsdata.rds&quot;</span>, <span class="at">package =</span> <span class="st">&quot;pavo&quot;</span>))</span>
<span id="cb99-2"><a href="analysing-data.html#cb99-2" aria-hidden="true" tabindex="-1"></a>mspecs <span class="ot">&lt;-</span> <span class="fu">aggspec</span>(specs, <span class="at">by =</span> <span class="dv">3</span>, <span class="at">FUN =</span> mean)</span>
<span id="cb99-3"><a href="analysing-data.html#cb99-3" aria-hidden="true" tabindex="-1"></a>spp <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">&#39;</span><span class="sc">\\</span><span class="st">.[0-9].*$&#39;</span>, <span class="st">&#39;&#39;</span>, <span class="fu">names</span>(mspecs))[<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb99-4"><a href="analysing-data.html#cb99-4" aria-hidden="true" tabindex="-1"></a>sppspec <span class="ot">&lt;-</span> <span class="fu">aggspec</span>(mspecs, <span class="at">by =</span> spp, <span class="at">FUN =</span> mean)</span>
<span id="cb99-5"><a href="analysing-data.html#cb99-5" aria-hidden="true" tabindex="-1"></a>spec.sm <span class="ot">&lt;-</span> <span class="fu">procspec</span>(sppspec, <span class="at">opt =</span> <span class="st">&quot;smooth&quot;</span>, <span class="at">span =</span> <span class="fl">0.2</span>)</span></code></pre></div>
<pre><code>## processing options applied:
## smoothing spectra with a span of 0.2</code></pre>
</div>
<div id="overview-1" class="section level3" number="3.1.2">
<h3><span class="header-section-number">3.1.2</span> Overview</h3>
<p><code>pavo</code> offers two main approaches for spectral data analysis. First, colour variables can be calculated based on the shape of the reflectance spectra. By using special <code>R</code> classes for spectra data frame objects, this can easily be done using the <code>summary</code> function with an <code>rspec</code> object (see below). The function <code>peakshape()</code> also returns descriptors for individual peaks in spectral curves, as outlined below.</p>
<p>Second, reflectance spectra can be analysed by accounting for the visual system receiving the colour signal, therefore representing reflectance spectra as perceived colours. To that end, we have implemented a suite of visual models and colourspaces including; the receptor-noise limited model of model of <span class="citation"><a href="#ref-Vorobyev1998" role="doc-biblioref">Vorobyev et al.</a> (<a href="#ref-Vorobyev1998" role="doc-biblioref">1998</a>)</span>, <span class="citation"><a href="#ref-Endler1990" role="doc-biblioref">Endler</a> (<a href="#ref-Endler1990" role="doc-biblioref">1990</a>)</span>’s segment classification method, flexible di- tri- and tetra-chromatic colourspaces, the Hexagon model of <span class="citation"><a href="#ref-Chittka1992" role="doc-biblioref">Chittka</a> (<a href="#ref-Chittka1992" role="doc-biblioref">1992</a>)</span>, the colour-opponent coding space of <span class="citation"><a href="#ref-Backhaus1991" role="doc-biblioref">Backhaus</a> (<a href="#ref-Backhaus1991" role="doc-biblioref">1991</a>)</span>, CIELAB and CIEXYZ spaces, and the categorical model of <span class="citation"><a href="#ref-Troje1993" role="doc-biblioref">Troje</a> (<a href="#ref-Troje1993" role="doc-biblioref">1993</a>)</span>.</p>
</div>
<div id="spectral-shape-analysis" class="section level3" number="3.1.3">
<h3><span class="header-section-number">3.1.3</span> Spectral Shape Analysis</h3>
<div id="colourimetric-variables" class="section level4" number="3.1.3.1">
<h4><span class="header-section-number">3.1.3.1</span> Colourimetric Variables</h4>
<p>Obtaining colourimetric variables (pertaining to hue, saturation and brightness/value) is pretty straightforward in <code>pavo</code>. Since reflectance spectra is stored in an object of class <code>rspec</code>, the <code>summary</code> function recognizes the object as such and extracts 23 variables, as outlined in <span class="citation"><a href="#ref-Montgomerie2006" role="doc-biblioref">Montgomerie</a> (<a href="#ref-Montgomerie2006" role="doc-biblioref">2006</a>)</span> (and reproduced <a href="spectral-shape-descriptors.html">here</a>). Though outlined in a book chapter on bird colouration, these variables are broadly applicable to any reflectance data, particularly if the taxon of interest has colour vision within the UV-human visible range.</p>
<p>The description and formulas for these variables can be found by running <code>?summary.rspec</code>.</p>
<div class="sourceCode" id="cb101"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb101-1"><a href="analysing-data.html#cb101-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(spec.sm)</span></code></pre></div>
<pre><code>##                B1       B2       B3        S1U        S1V        S1B       S1G
## cardinal 8984.266 22.40465 52.70167 0.16614848 0.19110215 0.11783711 0.1898843
## jacana   9668.503 24.11098 53.78744 0.09601072 0.11022142 0.19494363 0.3079239
## oriole   9108.474 22.71440 54.15508 0.07484472 0.08323604 0.05728179 0.3254691
## parakeet 6020.733 15.01430 29.86504 0.16792815 0.18501247 0.14407795 0.4244676
## robin    5741.392 14.31769 37.85542 0.08503645 0.10014331 0.14428130 0.2673730
## tanager  8515.251 21.23504 30.48108 0.20322408 0.23904955 0.26133545 0.3198724
##                S1Y       S1R        S2        S3         S4       S5       S6
## cardinal 0.2519174 0.5333116  7.607769 0.2954315 0.20287401 4003.465 45.77432
## jacana   0.2482692 0.4083129  7.101897 0.2471677 0.05315822 3314.024 46.21377
## oriole   0.3624909 0.5492239 14.507969 0.2986678 0.08730015 5233.013 50.42230
## parakeet 0.3395961 0.2718894  5.111074 0.4483808 0.27397999 1894.018 24.02183
## robin    0.2668138 0.5100635  9.130156 0.3031501         NA 2532.414 33.70923
## tanager  0.2430666 0.2238857  3.226417 0.3343114 0.16775097 1106.471 21.03373
##                  S7       S8          S9       S10  H1  H2  H3         H4  H5
## cardinal -0.2055325 2.043072  0.83791691 0.4144862 700 419 587 0.02130606 581
## jacana   -0.3054756 1.916710  0.73309578 0.1018889 700 593 529 0.71519543 468
## oriole   -0.5847404 2.219838  0.92483219 0.1937922 700 382 551 0.41436251 544
## parakeet  0.6656480 1.599931  0.58642273 0.4383490 572 618 634 0.97835569 506
## robin    -0.1076722 2.354377  0.81473611        NA 700  NA 593 0.42040115 631
## tanager  -0.7851970 0.990520 -0.04551277 0.1661607 557 594 362 1.52666571 518</code></pre>
<p><code>summary</code> also takes an additional argument <code>subset</code> which if changed from the default <code>FALSE</code> to <code>TRUE</code> will return only the most commonly used colourimetrics <span class="citation">(<a href="#ref-Andersson2006" role="doc-biblioref">Andersson and Prager 2006</a>)</span>. <code>summary</code> can also take a vector of colour variable names, which can be used to filter the results</p>
<div class="sourceCode" id="cb103"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb103-1"><a href="analysing-data.html#cb103-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(spec.sm, <span class="at">subset =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>##                B2       S8  H1
## cardinal 22.40465 2.043072 700
## jacana   24.11098 1.916710 700
## oriole   22.71440 2.219838 700
## parakeet 15.01430 1.599931 572
## robin    14.31769 2.354377 700
## tanager  21.23504 0.990520 557</code></pre>
<div class="sourceCode" id="cb105"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb105-1"><a href="analysing-data.html#cb105-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract only brightness variables</span></span>
<span id="cb105-2"><a href="analysing-data.html#cb105-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(spec.sm, <span class="at">subset =</span> <span class="fu">c</span>(<span class="st">&#39;B1&#39;</span>, <span class="st">&#39;B2&#39;</span>, <span class="st">&#39;B3&#39;</span>))</span></code></pre></div>
<pre><code>##                B1       B2       B3
## cardinal 8984.266 22.40465 52.70167
## jacana   9668.503 24.11098 53.78744
## oriole   9108.474 22.71440 54.15508
## parakeet 6020.733 15.01430 29.86504
## robin    5741.392 14.31769 37.85542
## tanager  8515.251 21.23504 30.48108</code></pre>
</div>
<div id="peak-shape-descriptors" class="section level4" number="3.1.3.2">
<h4><span class="header-section-number">3.1.3.2</span> Peak Shape Descriptors</h4>
<p>Particularly in cases of reflectance spectra that have multiple discrete peaks (in which case the <code>summary</code> function will only return variables based on the tallest peak in the curve), it might be useful to obtain variables that describe individual peak’s properties. The <code>peakshape()</code> function identifies the peak location (<code>H1</code>), returns the reflectance at that point (<code>B3</code>), and identifies the wavelengths at which the reflectance is half that at the peak, calculating the wavelength bandwidth of that interval (the Full Width at Half Maximum, or <code>FWHM</code>). The function also returns the half widths, which are useful when the peaks are located near the edge of the measurement limit and half maximum reflectance can only be reliably estimated from one of its sides.</p>
<p>If this all sounds too esoteric, fear not: <code>peakshape()</code> has the option of returning plots indicating what it’s calculating. The vertical continuous red line indicates the peak location, the horizontal continuous red line indicates the half-maximum reflectance, and the distance between the dashed lines (<code>HWHM.l</code> and <code>HWHM.r</code>) is the FWHM:</p>
<div class="sourceCode" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb107-1"><a href="analysing-data.html#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="fu">peakshape</span>(spec.sm, <span class="at">plot =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>##         id       B3  H1 FWHM HWHM.l HWHM.r incl.min
## 1 cardinal 52.70167 700   NA    114     NA      Yes
## 2   jacana 53.78744 700   NA    172     NA      Yes
## 3   oriole 54.15508 700   NA    149     NA      Yes
## 4 parakeet 29.86504 572  126     63     63      Yes
## 5    robin 37.85542 700   NA    107     NA      Yes
## 6  tanager 30.48108 557  166     81     85      Yes</code></pre>
<div class="figure"><span id="fig:unnamed-chunk-42"></span>
<img src="colsci-book_files/figure-html/unnamed-chunk-42-1.png" alt="Plots from `peakshape`" width="30%" /><img src="colsci-book_files/figure-html/unnamed-chunk-42-2.png" alt="Plots from `peakshape`" width="30%" /><img src="colsci-book_files/figure-html/unnamed-chunk-42-3.png" alt="Plots from `peakshape`" width="30%" /><img src="colsci-book_files/figure-html/unnamed-chunk-42-4.png" alt="Plots from `peakshape`" width="30%" /><img src="colsci-book_files/figure-html/unnamed-chunk-42-5.png" alt="Plots from `peakshape`" width="30%" /><img src="colsci-book_files/figure-html/unnamed-chunk-42-6.png" alt="Plots from `peakshape`" width="30%" />
<p class="caption">
Figure 3.1: Plots from <code>peakshape</code>
</p>
</div>
<p>As it can be seen, the variable FWHM is meaningless if the curve doesn’t have a clear peak. Sometimes, such as in the case of the Cardinal (Above figure, first panel), there might be a peak which is not the point of maximum reflectance of the entire spectral curve. <code>peakshape()</code> also offers a <code>select</code> argument to facilitate subsetting the spectra data frame to, for example, focus on a single reflectance peak:</p>
<div class="sourceCode" id="cb109"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb109-1"><a href="analysing-data.html#cb109-1" aria-hidden="true" tabindex="-1"></a><span class="fu">peakshape</span>(spec.sm, <span class="at">select =</span> <span class="dv">2</span>, <span class="at">lim =</span> <span class="fu">c</span>(<span class="dv">300</span>, <span class="dv">500</span>), <span class="at">plot =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-43"></span>
<img src="colsci-book_files/figure-html/unnamed-chunk-43-1.png" alt="Plot from `peakshape`, setting the wavelength limits to 300 and 500 nm" width="480" />
<p class="caption">
Figure 3.2: Plot from <code>peakshape</code>, setting the wavelength limits to 300 and 500 nm
</p>
</div>
<pre><code>##         id       B3  H1 FWHM HWHM.l HWHM.r incl.min
## 1 cardinal 17.84381 369  100     46     54      Yes</code></pre>
</div>
</div>
<div id="visual-modelling" class="section level3" number="3.1.4">
<h3><span class="header-section-number">3.1.4</span> Visual Modelling</h3>
<div id="overview-2" class="section level4" number="3.1.4.1">
<h4><span class="header-section-number">3.1.4.1</span> Overview</h4>
<p>All visual and colourspace modelling and plotting options are wrapped into four main function: <code>vismodel()</code>, <code>colspace()</code>, <code>coldist()</code> (or its bootstrap-based variant <code>bootcoldist()</code>), and <code>plot</code>. As detailed below, these functions cover the basic processes common to most modelling approaches. Namely, the estimation of quantum catches, their conversion to an appropriate space, estimating the distances between samples, and visualising the results. As a brief example, a typical workflow for modelling might use:</p>
<ol style="list-style-type: decimal">
<li><code>vismodel()</code> to estimate photoreceptor quantum catches. The assumptions of visual models may differ dramatically, so be sure to select options that are appropriate for your intended use.</li>
<li><code>colspace()</code> to convert the results of <code>vismodel()</code> (or user-provided quantum catches) into points in a given colourspace, specified with the <code>space</code> argument. If no <code>space</code> argument is provided, <code>colspace()</code> will automatically select the di-, tri- or tetrahedral colourspace, depending on the nature of the input data. The result of <code>colspace()</code> will be an object of class <code>colspace</code> (that inherits from <code>data.frame</code>), and will contain the location of stimuli in the selected space along with any associated colour variables.</li>
<li><code>coldist()</code> or <code>bootcoldist()</code> to estimate the colour distances between points in a colourspace (if the input is from <code>colspace</code>), or the ‘perceptual,’ noise-weighted distances in the receptor-noise limited model (if the input is from <code>vismodel()</code> where <code>relative = FALSE</code>).</li>
<li><code>plot</code> the output. <code>plot</code> will automatically select the appropriate visualisation based on the input <code>colspace</code> object, and will also accept various graphical parameters depending on the colourspace (see <code>?plot.colspace</code> and links therein for details).</li>
</ol>
</div>
<div id="accessing-and-plotting-in-built-spectral-data" class="section level4" number="3.1.4.2">
<h4><span class="header-section-number">3.1.4.2</span> Accessing and Plotting In-built Spectral Data</h4>
<p>All in-built spectral data in <code>pavo</code> can be easily retrieved and/or plotted via the <code>sensdata()</code> function. These data can be visualised directly by including <code>plot = TRUE</code>, or assigned to an object for subsequent use, as in:</p>
<div class="sourceCode" id="cb111"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb111-1"><a href="analysing-data.html#cb111-1" aria-hidden="true" tabindex="-1"></a>musca_sense <span class="ot">&lt;-</span> <span class="fu">sensdata</span>(<span class="at">visual =</span> <span class="st">&quot;musca&quot;</span>, <span class="at">achromatic =</span> <span class="st">&quot;md.r1&quot;</span>)</span>
<span id="cb111-2"><a href="analysing-data.html#cb111-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(musca_sense)</span></code></pre></div>
<pre><code>##    wl    musca.u     musca.s     musca.m     musca.l       md.r1
## 1 300 0.01148992 0.006726680 0.001533875 0.001524705 0.001713692
## 2 301 0.01152860 0.006749516 0.001535080 0.001534918 0.001731120
## 3 302 0.01156790 0.006773283 0.001536451 0.001545578 0.001751792
## 4 303 0.01160839 0.006798036 0.001537982 0.001556695 0.001775960
## 5 304 0.01165060 0.006823827 0.001539667 0.001568283 0.001803871
## 6 305 0.01169510 0.006850712 0.001541499 0.001580352 0.001835777</code></pre>
</div>
<div id="visual-phenotypes" class="section level4" number="3.1.4.3">
<h4><span class="header-section-number">3.1.4.3</span> Visual Phenotypes</h4>
<p><code>pavo</code> contains numerous di-, tri- and tetrachromatic visual systems, to accompany the suite of new models. The full complement of included systems are accessible via the <code>vismodel()</code> argument <code>visual</code>, and include:</p>
<table>
<caption><span id="tab:unnamed-chunk-45">Table 3.1: </span>Built-in visual phenotypes available in pavo</caption>
<colgroup>
<col width="10%" />
<col width="89%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">phenotype</th>
<th align="left">description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">avg.uv</td>
<td align="left">average ultraviolet-sensitive avian (tetrachromat)</td>
</tr>
<tr class="even">
<td align="left">avg.v</td>
<td align="left">average violet-sensitive avian (tetrachromat)</td>
</tr>
<tr class="odd">
<td align="left">bluetit</td>
<td align="left">The blue tit <em>Cyanistes caeruleus</em> (tetrachromat)</td>
</tr>
<tr class="even">
<td align="left">star</td>
<td align="left">The starling <em>Sturnus vulgaris</em> (tetrachromat)</td>
</tr>
<tr class="odd">
<td align="left">pfowl</td>
<td align="left">The peafowl <em>Pavo cristatus</em> (tetrachromat)</td>
</tr>
<tr class="even">
<td align="left">apis</td>
<td align="left">The honeybee <em>Apis mellifera</em> (trichromat)</td>
</tr>
<tr class="odd">
<td align="left">ctenophorus</td>
<td align="left">The ornate dragon lizard <em>Ctenophorus ornatus</em> (trichromat)</td>
</tr>
<tr class="even">
<td align="left">canis</td>
<td align="left">The canid <em>Canis familiaris</em> (dichromat)</td>
</tr>
<tr class="odd">
<td align="left">musca</td>
<td align="left">The housefly <em>Musca domestica</em> (tetrachromat)</td>
</tr>
<tr class="even">
<td align="left">cie2</td>
<td align="left">2-degree colour matching functions for CIE models of human colour vision (trichromat)</td>
</tr>
<tr class="odd">
<td align="left">cie10</td>
<td align="left">10-degree colour matching functions for CIE models of human colour vision (trichromat)</td>
</tr>
<tr class="even">
<td align="left">segment</td>
<td align="left">A generic ‘viewer’ with broad sensitivities for use in the segment analysis of Endler (1990) (tetrachromat)</td>
</tr>
<tr class="odd">
<td align="left">habronattus</td>
<td align="left">The jumping spider <em>Habronattus pyrrithrix</em> (trichromat)</td>
</tr>
<tr class="even">
<td align="left">rhinecanthus</td>
<td align="left">The triggerfish <em>Rhinecanthus aculeatus</em> (trichromat)</td>
</tr>
</tbody>
</table>
</div>
<div id="estimating-quantum-catch" class="section level4" number="3.1.4.4">
<h4><span class="header-section-number">3.1.4.4</span> Estimating Quantum Catch</h4>
<p>Numerous models have been developed to understand how colours are perceived and discriminated by an individual’s or species’ visual system <span class="citation">(described in detail in <a href="#ref-Endler1990" role="doc-biblioref">Endler 1990</a>; <a href="#ref-Renoult2017" role="doc-biblioref">Renoult, Kelber, and Schaefer 2017</a>; <a href="#ref-Vorobyev1998" role="doc-biblioref">Vorobyev et al. 1998</a>)</span>. In essence, these models take into account the receptor sensitivity of the different receptors that make the visual system in question and quantify how a given colour would stimulate those receptors individually, and their combined effect on the perception of colour. These models also have an important component of assuming and interpreting the chromatic component of colour (hue and saturation) to be processed independently of the achromatic (brightness, or luminance) component. This provides a flexible framework allowing for a tiered model construction, in which information on aspects such as different illumination sources, backgrounds, and visual systems can be considered and compared.</p>
<p>To apply any such model, we first need to quantify receptor excitation and then consider how the signal is being processed, while possibly considering the relative density of different cones and the noise-to-signal ratio.</p>
<p>To quantify the stimulation of receptors by a given stimulus, we will use the function <code>vismodel()</code>. This takes an <code>rspec</code> dataframe as a minimal input, and the user can either select from the available options or input their own data for the additional arguments in the function:</p>
<ul>
<li><code>visual</code>: the visual system to be used. Available inbuilt options are detailed above, or the user may include their own dataframe, with the first column being the wavelength range and the following columns being the absorbance at each wavelength for each cone type (see below for an example).</li>
<li><code>achromatic</code>: Either a receptor’s sensitivity data (available options include the blue tit, chicken, and starling double-cones, and the housefly’s R1-6 receptor), which can also be user-defined as above; or the longest-wavelength receptor, the sum of the two longest-wavelength receptors, or the sum of all receptors can be used. Alternatively, <code>none</code> can be specified for no achromatic stimulation calculation.</li>
<li><code>illum</code>: The illuminant being considered. By default, it considers an ideal white illuminant, but implemented options are a blue sky, standard daylight, and forest shade illuminants. A vector of same length as the wavelength range being considered can also be used as the input.</li>
<li><code>trans</code>: Models of the effects of light transmission (e.g. through noisy environments or ocular filters). The argument defaults to <code>ideal</code> (i.e. no effect), though users can also use the built-in options of <code>bluetit</code> or <code>blackbird</code> to model the ocular transmission of blue tits/blackbirds, or specify a user-defined vector containing transmission spectra.</li>
<li><code>qcatch</code>: This argument determines what photon catch data should be returned
<ul>
<li><code>Qi</code>: The receptor quantum catches, calculated for receptor <span class="math inline">\(i\)</span> as:
<span class="math display">\[Q_i = \int_\lambda{R_i(\lambda)S(\lambda)I(\lambda)d\lambda}\]</span>
Where <span class="math inline">\(\lambda\)</span> denotes the wavelength, <span class="math inline">\(R_i(\lambda)\)</span> the spectral sensitivity of receptor <span class="math inline">\(i\)</span>, <span class="math inline">\(S(\lambda)\)</span> the reflectance spectrum of the colour, and <span class="math inline">\(I(\lambda)\)</span> the illuminant spectrum.</li>
<li><code>fi</code>: The receptor quantum catches transformed according to Fechner’s law, in which the signal of the receptor is proportional to the logarithm of the quantum catch i.e. <span class="math inline">\(f_i = \ln(Q_i)\)</span></li>
<li><code>Ei</code>: the hyperbolic transform (a simplification of the Michaelis–Menten photoreceptor equation), where <span class="math display">\[E_i = \frac{Q_i}{Q_i + 1}\]</span>.</li>
</ul></li>
<li><code>bkg</code>: The background being considered. By default, it considers an idealized background (i.e. wavelength-independent influence of the background on colour). A vector of same length as the wavelength range being considered can also be used as the input.</li>
<li><code>vonkries</code>: a logical argument which determines if the von Kries transformation (which normalizes receptor quantum catches to the background, thus accounting for receptor adaptation) is to be applied (defaults to <code>FALSE</code>). If <code>TRUE</code>, <span class="math inline">\(Q_i\)</span> is multiplied by a constant <span class="math inline">\(k\)</span>, which describes the von Kries transformation: <span class="math display">\[k_i = \frac{1}{\int_\lambda R_i(\lambda)S^b(\lambda)I(\lambda)d\lambda}\]</span> Where <span class="math inline">\(S^b\)</span> denotes the reflectance spectra of the background.</li>
<li><code>scale</code>: This argument defines how the illuminant should be scaled. The scale of the illuminant is critical for receptor noise models in which the signal intensity influences the noise (see Receptor noise section, below). Illuminant curves should be in units of <span class="math inline">\(\mu mol.s^{-1}.m^{-2}\)</span> in order to yield physiologically meaningful results. (Some software return illuminant information values in <span class="math inline">\(\mu Watt.cm^{-2}\)</span>, and must be converted to <span class="math inline">\(\mu mol.s^{-1}.m^{-2}\)</span>. This can be done by using the <code>irrad2flux()</code> and <code>flux2irrad()</code> functions.) Therefore, if the user-specified illuminant curves are <em>not</em> in these units (i.e. are measured proportional to a white standard, for example), the <code>scale</code> parameter can be used as a multiplier to yield curves that are at least a reasonable approximation of the illuminant value. Commonly used values are <strong>500</strong> for dim conditions and <strong>10,000</strong> for bright conditions.</li>
<li><code>relative</code>: If <code>TRUE</code>, it will make the cone stimulations relative to their sum. This is appropriate for general colourspace models<span class="citation">(e.g. <a href="#ref-Goldsmith1990" role="doc-biblioref">Goldsmith 1990</a>; <a href="#ref-Stoddard2008" role="doc-biblioref">Stoddard and Prum 2008</a>)</span>. For the receptor-noise model, however, it is important to set <code>relative = FALSE</code>.</li>
</ul>
<p>All visual models begin with the estimation of receptor quantum catches. The requirements of models may differ significantly of course, so be sure to consult the function documentation and original publications. For this example, we will use the average reflectance of the different species to calculate the raw stimulation of retinal cones, considering the avian average UV visual system, a standard daylight illumination, and an idealized background.</p>
<div class="sourceCode" id="cb113"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb113-1"><a href="analysing-data.html#cb113-1" aria-hidden="true" tabindex="-1"></a>vismod1 <span class="ot">&lt;-</span> <span class="fu">vismodel</span>(sppspec,</span>
<span id="cb113-2"><a href="analysing-data.html#cb113-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">visual =</span> <span class="st">&quot;avg.uv&quot;</span>, <span class="at">achromatic =</span> <span class="st">&quot;bt.dc&quot;</span>,</span>
<span id="cb113-3"><a href="analysing-data.html#cb113-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">illum =</span> <span class="st">&quot;D65&quot;</span>, <span class="at">relative =</span> <span class="cn">FALSE</span></span>
<span id="cb113-4"><a href="analysing-data.html#cb113-4" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>Since there are multiple parameters that can be used to customize the output of <code>vismodel()</code>, as detailed above, for convenience these can be returned by using <code>summary</code> in a <code>vismodel</code> object:</p>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb114-1"><a href="analysing-data.html#cb114-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(vismod1)</span></code></pre></div>
<pre><code>## visual model options:
##  * Quantal catch: Qi 
##  * Visual system, chromatic: avg.uv 
##  * Visual system, achromatic: bt.dc 
##  * Illuminant: D65, scale = 1 (von Kries colour correction not applied) 
##  * Background: ideal 
##  * Transmission: ideal 
##  * Relative: FALSE</code></pre>
<pre><code>##        u                 s                 m                l         
##  Min.   :0.01671   Min.   :0.04199   Min.   :0.1325   Min.   :0.2092  
##  1st Qu.:0.02408   1st Qu.:0.06860   1st Qu.:0.1753   1st Qu.:0.2232  
##  Median :0.03039   Median :0.07148   Median :0.2675   Median :0.2759  
##  Mean   :0.03599   Mean   :0.09924   Mean   :0.2338   Mean   :0.3076  
##  3rd Qu.:0.04731   3rd Qu.:0.14265   3rd Qu.:0.2819   3rd Qu.:0.3815  
##  Max.   :0.06353   Max.   :0.17648   Max.   :0.3037   Max.   :0.4625  
##       lum        
##  Min.   :0.1513  
##  1st Qu.:0.1966  
##  Median :0.2247  
##  Mean   :0.2241  
##  3rd Qu.:0.2661  
##  Max.   :0.2766</code></pre>
<p>We can visualise what <code>vismodel()</code> is doing when estimating quantum catches by comparing the reflectance spectra to the estimates they are generating:</p>
<div class="sourceCode" id="cb117"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb117-1"><a href="analysing-data.html#cb117-1" aria-hidden="true" tabindex="-1"></a>oldpar <span class="ot">&lt;-</span> <span class="fu">par</span>(<span class="at">no.readonly =</span> <span class="cn">TRUE</span>)</span>
<span id="cb117-2"><a href="analysing-data.html#cb117-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">6</span>), <span class="at">oma =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">3</span>, <span class="dv">0</span>, <span class="dv">0</span>))</span>
<span id="cb117-3"><a href="analysing-data.html#cb117-3" aria-hidden="true" tabindex="-1"></a><span class="fu">layout</span>(<span class="fu">rbind</span>(<span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">4</span> , <span class="dv">3</span>, <span class="dv">6</span> , <span class="dv">5</span>),</span>
<span id="cb117-4"><a href="analysing-data.html#cb117-4" aria-hidden="true" tabindex="-1"></a>             <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">3</span> , <span class="dv">3</span>, <span class="dv">5</span> , <span class="dv">5</span>), </span>
<span id="cb117-5"><a href="analysing-data.html#cb117-5" aria-hidden="true" tabindex="-1"></a>             <span class="fu">c</span>(<span class="dv">8</span>, <span class="dv">7</span>, <span class="dv">10</span>, <span class="dv">9</span>, <span class="dv">12</span>, <span class="dv">11</span>),</span>
<span id="cb117-6"><a href="analysing-data.html#cb117-6" aria-hidden="true" tabindex="-1"></a>             <span class="fu">c</span>(<span class="dv">7</span>, <span class="dv">7</span>, <span class="dv">9</span> , <span class="dv">9</span>, <span class="dv">11</span>, <span class="dv">11</span>)))</span>
<span id="cb117-7"><a href="analysing-data.html#cb117-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-8"><a href="analysing-data.html#cb117-8" aria-hidden="true" tabindex="-1"></a>sppspecol <span class="ot">&lt;-</span> <span class="fu">spec2rgb</span>(sppspec)</span>
<span id="cb117-9"><a href="analysing-data.html#cb117-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-10"><a href="analysing-data.html#cb117-10" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>) {</span>
<span id="cb117-11"><a href="analysing-data.html#cb117-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">par</span>(<span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>))</span>
<span id="cb117-12"><a href="analysing-data.html#cb117-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot</span>(sppspec, <span class="at">select =</span> i<span class="sc">+</span><span class="dv">1</span>, <span class="at">col =</span> sppspecol, <span class="at">lwd =</span> <span class="dv">3</span>, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">100</span>))</span>
<span id="cb117-13"><a href="analysing-data.html#cb117-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">par</span>(<span class="at">mar =</span> <span class="fu">c</span>(<span class="fl">4.1</span>, <span class="fl">2.5</span>, <span class="fl">2.5</span>, <span class="dv">2</span>))</span>
<span id="cb117-14"><a href="analysing-data.html#cb117-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">barplot</span>(<span class="fu">as.matrix</span>(vismod1[i, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>]), <span class="at">yaxt =</span> <span class="st">&quot;n&quot;</span>, <span class="at">col =</span> <span class="st">&quot;black&quot;</span>)</span>
<span id="cb117-15"><a href="analysing-data.html#cb117-15" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb117-16"><a href="analysing-data.html#cb117-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-17"><a href="analysing-data.html#cb117-17" aria-hidden="true" tabindex="-1"></a><span class="fu">mtext</span>(<span class="st">&quot;Wavelength (nm)&quot;</span>, <span class="at">side =</span> <span class="dv">1</span>, <span class="at">outer =</span> <span class="cn">TRUE</span>, <span class="at">line =</span> <span class="dv">1</span>)</span>
<span id="cb117-18"><a href="analysing-data.html#cb117-18" aria-hidden="true" tabindex="-1"></a><span class="fu">mtext</span>(<span class="st">&quot;Reflectance (%)&quot;</span>, <span class="at">side =</span> <span class="dv">2</span>, <span class="at">outer =</span> <span class="cn">TRUE</span>, <span class="at">line =</span> <span class="dv">1</span>)</span></code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-48"></span>
<img src="colsci-book_files/figure-html/unnamed-chunk-48-1.png" alt="Plots of species mean reflectance curves with corresponding relative usml cone stimulations (insets)." width="816" />
<p class="caption">
Figure 3.3: Plots of species mean reflectance curves with corresponding relative usml cone stimulations (insets).
</p>
</div>
<div class="sourceCode" id="cb118"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb118-1"><a href="analysing-data.html#cb118-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(oldpar)</span></code></pre></div>
<p>As described above, <code>vismodel</code> also accepts user-defined visual systems, background and illuminants. We will illustrate this by showcasing the function <code>sensmodel</code>, which models spectral sensitivities of retinas based on their peak cone sensitivity, as described in <span class="citation"><a href="#ref-Govardovskii2000" role="doc-biblioref">Govardovskii et al.</a> (<a href="#ref-Govardovskii2000" role="doc-biblioref">2000</a>)</span> and <span class="citation"><a href="#ref-Hart2005" role="doc-biblioref">Hart and Vorobyev</a> (<a href="#ref-Hart2005" role="doc-biblioref">2005</a>)</span>. <code>sensmodel</code> takes several optional arguments, but the main one is a vector containing the peak sensitivities for the cones being modelled. Let’s model an idealized dichromat visual system, with cones peaking in sensitivity at 350 and 650 nm:</p>
<div class="sourceCode" id="cb119"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb119-1"><a href="analysing-data.html#cb119-1" aria-hidden="true" tabindex="-1"></a>idealizeddichromat <span class="ot">&lt;-</span> <span class="fu">sensmodel</span>(<span class="fu">c</span>(<span class="dv">350</span>, <span class="dv">650</span>))</span>
<span id="cb119-2"><a href="analysing-data.html#cb119-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(idealizeddichromat, <span class="at">col =</span> <span class="fu">spec2rgb</span>(idealizeddichromat), <span class="at">ylab =</span> <span class="st">&quot;Absorbance&quot;</span>)</span></code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-49"></span>
<img src="colsci-book_files/figure-html/unnamed-chunk-49-1.png" alt="Idealized dichromat photoreceptors created using `sensmodel`." width="480" />
<p class="caption">
Figure 3.4: Idealized dichromat photoreceptors created using <code>sensmodel</code>.
</p>
</div>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="analysing-data.html#cb120-1" aria-hidden="true" tabindex="-1"></a>vismod.idi <span class="ot">&lt;-</span> <span class="fu">vismodel</span>(sppspec, <span class="at">visual =</span> idealizeddichromat, <span class="at">relative =</span> <span class="cn">FALSE</span>)</span>
<span id="cb120-2"><a href="analysing-data.html#cb120-2" aria-hidden="true" tabindex="-1"></a>vismod.idi</span></code></pre></div>
<pre><code>##             lmax350   lmax650 lum
## cardinal 0.14578161 0.3519237  NA
## jacana   0.09155090 0.3179246  NA
## oriole   0.06981615 0.3775647  NA
## parakeet 0.10579320 0.1727374  NA
## robin    0.04727061 0.2174901  NA
## tanager  0.16438304 0.2148739  NA</code></pre>
</div>
<div id="the-receptor-noise-model" class="section level4" number="3.1.4.5">
<h4><span class="header-section-number">3.1.4.5</span> The Receptor Noise Model</h4>
<p>The receptor-noise limited (RNL) model of <span class="citation"><a href="#ref-Vorobyev1998" role="doc-biblioref">Vorobyev et al.</a> (<a href="#ref-Vorobyev1998" role="doc-biblioref">1998</a>)</span>, <span class="citation"><a href="#ref-Vorobyev2001" role="doc-biblioref">Vorobyev et al.</a> (<a href="#ref-Vorobyev2001" role="doc-biblioref">2001</a>)</span> offers a basis for estimating the ‘perceptual’ distance between coloured stimuli (depending on available physiological and behavioural data), and assumes that the simultaneous discrimination of colours is fundamentally limited by photoreceptor noise. Colour distances under the RNL model can be calculated by using the inverse of the noise-to-signal ratio, known as the Weber fraction (<span class="math inline">\(w_i\)</span> for each cone <span class="math inline">\(i\)</span>). The Weber fraction can be calculated from the noise-to-signal ratio of cone <span class="math inline">\(i\)</span> (<span class="math inline">\(v_i\)</span>) and the relative number of receptor cells of type <span class="math inline">\(i\)</span> within the receptor field (<span class="math inline">\(n_i\)</span>):</p>
<p><span class="math display">\[w_i = \frac{v_i}{\sqrt{n_i}}\]</span></p>
<p><span class="math inline">\(w_i\)</span> is the value used for the noise when considering only neural noise mechanisms. Alternatively, the model can consider that the intensity of the colour signal itself contributes to the noise (photoreceptor, or quantum, noise). In this case, the noise for a receptor <span class="math inline">\(i\)</span> is calculated as:</p>
<p><span class="math display">\[w_i = \sqrt{\frac{v_i^2}{\sqrt{n_i}} + \frac{2}{Q_a+Q_b}}\]</span></p>
<p>where <span class="math inline">\(a\)</span> and <span class="math inline">\(b\)</span> refer to the two colour signals being compared. Note that when the values of <span class="math inline">\(Q_a\)</span> and <span class="math inline">\(Q_b\)</span> are very high, the second portion of the equation tends to zero, and the both formulas should yield similar results. Hence, it is important that the quantum catch are calculated in the appropriate illuminant scale, as described above.</p>
<p>Colour distances are obtained by weighting the Euclidean distance of the photoreceptor quantum catches by the Weber fraction of the cones (<span class="math inline">\(\Delta S\)</span>). These measurements are in units of Just Noticeable Differences (JNDs), where distances over a certain threshold (usually 1) are considered to be discernible under the conditions considered (e.g., backgrounds, illumination). The equations used in these calculations are:</p>
<p>For dichromats:
<span class="math display">\[\Delta S = \sqrt{\frac{(\Delta f_1 - \Delta f_2)^2}{w_1^2+w_2^2}}\]</span></p>
<p>For trichromats:
<span class="math display">\[\Delta S = \sqrt{\frac{w_1^2(\Delta f_3 - \Delta f_2)^2 + w_2^2(\Delta f_3 - \Delta f_1)^2 +
w_3^2(\Delta f_1 - \Delta f_2)^2 }{ (w_1w_2)^2 + (w_1w_3)^2 + (w_2w_3)^2 }}\]</span></p>
<p>For tetrachromats:
<span class="math display">\[\Delta S =
\sqrt{(w_1w_2)^2(\Delta f_4 - \Delta f_3)^2 + (w_1w_3)^2(\Delta f_4 - \Delta f_2)^2 +
(w_1w_4)^2(\Delta f_3 - \Delta f_2)^2 + \\ (w_2w_3)^2(\Delta f_4 - \Delta f_1)^2 +
(w_2w_4)^2(\Delta f_3 - \Delta f_1)^2 + (w_3w_4)^2(\Delta f_2 - \Delta f_1)^2 / \\
((w_1w_2w_3)^2 + (w_1w_2w_4)^2 + (w_1w_3w_4)^2 + (w_2w_3w_4)^2)}\]</span></p>
<p>For the chromatic contrast. The achromatic contrast (<span class="math inline">\(\Delta L\)</span>) can be calculated based on the double cone or the receptor (or combination of receptors) responsible for chromatic processing by the equation:</p>
<p><span class="math display">\[\Delta L = \frac{\Delta f}{w}\]</span></p>
<div id="noise-weighted-distances-with-coldist-and-bootcoldist" class="section level5" number="3.1.4.5.1">
<h5><span class="header-section-number">3.1.4.5.1</span> Noise-weighted Distances with <code>coldist()</code> and <code>bootcoldist()</code></h5>
<p><code>pavo</code> implements the noise-weighted colour distance calculations in the functions <code>coldist()</code> and <code>bootcoldist()</code>, assuming that raw receptor quantum catches are provided (via <code>relative = FALSE</code> in <code>vismodel()</code>). For the achromatic contrast, <code>coldist()</code> uses <code>n4</code> to calculate <span class="math inline">\(w\)</span> for the achromatic contrast. Note that even if <span class="math inline">\(Q_i\)</span> is chosen, values are still log-transformed. This option is available in case the user wants to specify a data frame of quantum catches that was not generated by <code>vismodel()</code> as an input. In this case, the argument <code>qcatch</code> should be used to inform the function if <span class="math inline">\(Q_i\)</span> or <span class="math inline">\(f_i\)</span> values are being used (note that if the input to <code>coldist()</code> is an object generated using the <code>vismodel()</code> function, this argument is ignored.) The type of noise to be calculated can be selected from the <code>coldist()</code> argument <code>noise</code> (which accepts either <code>"neural"</code> or <code>"quantum"</code>).</p>
<div class="sourceCode" id="cb122"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb122-1"><a href="analysing-data.html#cb122-1" aria-hidden="true" tabindex="-1"></a><span class="fu">coldist</span>(vismod1,</span>
<span id="cb122-2"><a href="analysing-data.html#cb122-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">noise =</span> <span class="st">&quot;neural&quot;</span>, <span class="at">achromatic =</span> <span class="cn">TRUE</span>, <span class="at">n =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">4</span>),</span>
<span id="cb122-3"><a href="analysing-data.html#cb122-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">weber =</span> <span class="fl">0.1</span>, <span class="at">weber.achro =</span> <span class="fl">0.1</span></span>
<span id="cb122-4"><a href="analysing-data.html#cb122-4" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>##      patch1   patch2        dS        dL
## 1  cardinal   jacana  8.436191 3.6082787
## 2  cardinal   oriole  7.874540 3.5245174
## 3  cardinal parakeet  8.010167 0.7553289
## 4  cardinal    robin  5.865464 2.4275512
## 5  cardinal  tanager 10.116173 2.2442493
## 6    jacana   oriole 10.114671 0.0837613
## 7    jacana parakeet  4.316712 2.8529498
## 8    jacana    robin  3.499380 6.0358299
## 9    jacana  tanager  5.064380 1.3640294
## 10   oriole parakeet  8.147778 2.7691885
## 11   oriole    robin  7.200079 5.9520686
## 12   oriole  tanager 13.532690 1.2802681
## 13 parakeet    robin  4.744364 3.1828801
## 14 parakeet  tanager  5.886396 1.4889204
## 15    robin  tanager  7.679779 4.6718005</code></pre>
<div class="sourceCode" id="cb124"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb124-1"><a href="analysing-data.html#cb124-1" aria-hidden="true" tabindex="-1"></a><span class="fu">coldist</span>(vismod.idi, <span class="at">n =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">weber =</span> <span class="fl">0.1</span>)</span></code></pre></div>
<pre><code>##      patch1   patch2        dS dL
## 1  cardinal   jacana 2.0993283 NA
## 2  cardinal   oriole 4.6567462 NA
## 3  cardinal parakeet 2.2575457 NA
## 4  cardinal    robin 3.7236780 NA
## 5  cardinal  tanager 3.5417700 NA
## 6    jacana   oriole 2.5574178 NA
## 7    jacana parakeet 4.3568740 NA
## 8    jacana    robin 1.6243496 NA
## 9    jacana  tanager 5.6410983 NA
## 10   oriole parakeet 6.9142918 NA
## 11   oriole    robin 0.9330682 NA
## 12   oriole  tanager 8.1985162 NA
## 13 parakeet    robin 5.9812236 NA
## 14 parakeet  tanager 1.2842244 NA
## 15    robin  tanager 7.2654480 NA</code></pre>
<p>Where <code>dS</code> is the chromatic contrast (<span class="math inline">\(\Delta S\)</span>) and <code>dL</code> is the achromatic contrast (<span class="math inline">\(\Delta L\)</span>). Note that, by default, <code>achromatic = FALSE</code>, so <code>dL</code> isn’t included in the second result (this is to maintain consistency since, in the <code>vismodel()</code> function, the <code>achromatic</code> argument defaults to <code>none</code>). As expected, values are really high under the avian colour vision, since the colours of these species are quite different and because of the enhanced discriminatory ability with four compared to two cones.</p>
<p><code>coldist()</code> also has a <code>subset</code> argument, which is useful if only certain comparisons are of interest (for example, of colour patches against a background, or only comparisons among a species or body patch). <code>subset</code> can be a vector of length one or two. If only one subsetting option is passed, all comparisons against the matching argument are returned (useful in the case of comparing to a background, for example). If two values are passed, comparisons will only be made between samples that match that rule (partial string matching and regular expressions are accepted). For example, compare:</p>
<div class="sourceCode" id="cb126"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb126-1"><a href="analysing-data.html#cb126-1" aria-hidden="true" tabindex="-1"></a><span class="fu">coldist</span>(vismod1, <span class="at">subset =</span> <span class="st">&#39;cardinal&#39;</span>)</span></code></pre></div>
<pre><code>##     patch1   patch2        dS dL
## 1 cardinal   jacana  8.436191 NA
## 2 cardinal   oriole  7.874540 NA
## 3 cardinal parakeet  8.010167 NA
## 4 cardinal    robin  5.865464 NA
## 5 cardinal  tanager 10.116173 NA</code></pre>
<p>to:</p>
<div class="sourceCode" id="cb128"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb128-1"><a href="analysing-data.html#cb128-1" aria-hidden="true" tabindex="-1"></a><span class="fu">coldist</span>(vismod1, <span class="at">subset =</span> <span class="fu">c</span>(<span class="st">&#39;cardinal&#39;</span>, <span class="st">&#39;jacana&#39;</span>))</span></code></pre></div>
<pre><code>##     patch1 patch2       dS dL
## 1 cardinal jacana 8.436191 NA</code></pre>
<p>A bootstrap-based variant of <code>coldist()</code>, described in <span class="citation"><a href="#ref-Maia2018" role="doc-biblioref">Maia and White</a> (<a href="#ref-Maia2018" role="doc-biblioref">2018</a>)</span> as part of a more general analytical framework, is implemented via the function <code>bootcoldist()</code>. As the name suggests, <code>bootcoldist()</code> uses a bootstrap procedure to generate confidence intervals for the mean chromatic and/or achromatic distance between two or more samples of colours. This has the benefit of generating useful errors around distance estimates, which can be inspected for the inclusion of appropriate ‘threshold,’ when asking common kinds of questions that relate to between-group differences (e.g. animals-vs-backgrounds, sexual dichromatism, colour polymorphism). The function’s arguments are shared <code>coldist()</code> save for a few extra requirements. The <code>by</code> argument is used to specify a grouping variable that defines the groups to be compared, while <code>boot.n</code> and <code>alpha</code> control the number of bootstrap replicates and the confidence level for the confidence interval, respectively. The latter two have sensible defaults, but the <code>by</code> argument is required.</p>
<p>Here we run present a quick example that tests for colour differences across three body regions (breast, crown, and throat) of the finch <em>Sicalis citrina</em>, given seven measures of each region from different males.</p>
<div class="sourceCode" id="cb130"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb130-1"><a href="analysing-data.html#cb130-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the data</span></span>
<span id="cb130-2"><a href="analysing-data.html#cb130-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(sicalis)</span>
<span id="cb130-3"><a href="analysing-data.html#cb130-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb130-4"><a href="analysing-data.html#cb130-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Construct a model using an avian viewer</span></span>
<span id="cb130-5"><a href="analysing-data.html#cb130-5" aria-hidden="true" tabindex="-1"></a>sicmod <span class="ot">&lt;-</span> <span class="fu">vismodel</span>(sicalis, <span class="at">visual =</span> <span class="st">&#39;avg.uv&#39;</span>, <span class="at">relative =</span> <span class="cn">FALSE</span>)</span>
<span id="cb130-6"><a href="analysing-data.html#cb130-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb130-7"><a href="analysing-data.html#cb130-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a grouping variable to group by body region,</span></span>
<span id="cb130-8"><a href="analysing-data.html#cb130-8" aria-hidden="true" tabindex="-1"></a><span class="co"># informed by the spectral data names</span></span>
<span id="cb130-9"><a href="analysing-data.html#cb130-9" aria-hidden="true" tabindex="-1"></a>regions <span class="ot">&lt;-</span> <span class="fu">substr</span>(<span class="fu">rownames</span>(sicmod), <span class="dv">6</span>, <span class="dv">6</span>)</span>
<span id="cb130-10"><a href="analysing-data.html#cb130-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb130-11"><a href="analysing-data.html#cb130-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Estimate bootstrapped colour-distances</span></span>
<span id="cb130-12"><a href="analysing-data.html#cb130-12" aria-hidden="true" tabindex="-1"></a>sicdist <span class="ot">&lt;-</span> <span class="fu">bootcoldist</span>(sicmod, <span class="at">by =</span> regions, <span class="at">n =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">4</span>), <span class="at">weber =</span> <span class="fl">0.05</span>)</span>
<span id="cb130-13"><a href="analysing-data.html#cb130-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb130-14"><a href="analysing-data.html#cb130-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a look at the resulting pairwise estimates </span></span>
<span id="cb130-15"><a href="analysing-data.html#cb130-15" aria-hidden="true" tabindex="-1"></a>sicdist</span></code></pre></div>
<pre><code>##       dS.mean    dS.lwr    dS.upr
## B-C  9.253097 5.7015179 14.013095
## B-T  3.483528 0.3728945  9.509223
## C-T 12.221038 8.1736179 16.899949</code></pre>
<p>And plot the results for good measure. We can see that the pairwise comparisons of the chest, breast, and throat are quite chromatically distinct, but the breast and throat are much more similar in colour, perhaps to the point of being indistinguishable.</p>
<div class="sourceCode" id="cb132"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb132-1"><a href="analysing-data.html#cb132-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sicdist[, <span class="dv">1</span>], </span>
<span id="cb132-2"><a href="analysing-data.html#cb132-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">20</span>), </span>
<span id="cb132-3"><a href="analysing-data.html#cb132-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">pch =</span> <span class="dv">21</span>, </span>
<span id="cb132-4"><a href="analysing-data.html#cb132-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">bg =</span> <span class="dv">1</span>, </span>
<span id="cb132-5"><a href="analysing-data.html#cb132-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">cex =</span> <span class="dv">2</span>, </span>
<span id="cb132-6"><a href="analysing-data.html#cb132-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">xaxt =</span> <span class="st">&#39;n&#39;</span>, </span>
<span id="cb132-7"><a href="analysing-data.html#cb132-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">&#39;Centroid comparison&#39;</span>, </span>
<span id="cb132-8"><a href="analysing-data.html#cb132-8" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">&#39;Chromatic contrast (dS)&#39;</span>)</span>
<span id="cb132-9"><a href="analysing-data.html#cb132-9" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="dv">1</span>, <span class="at">at =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="at">labels =</span> <span class="fu">rownames</span>(sicdist))</span>
<span id="cb132-10"><a href="analysing-data.html#cb132-10" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, sicdist[, <span class="dv">2</span>], <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, sicdist[, <span class="dv">3</span>], <span class="at">lwd =</span> <span class="dv">2</span>)  <span class="co"># Add CI&#39;s</span></span>
<span id="cb132-11"><a href="analysing-data.html#cb132-11" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h =</span> <span class="dv">1</span>, <span class="at">lty =</span> <span class="dv">3</span>, <span class="at">lwd =</span> <span class="dv">2</span>)  <span class="co"># Add a &#39;threshold&#39; line at dS = 1</span></span></code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-55"></span>
<img src="colsci-book_files/figure-html/unnamed-chunk-55-1.png" alt="The bootstrapped colour distances between the throat, chest, and breast regions of male Sicalis citrina." width="480" />
<p class="caption">
Figure 3.5: The bootstrapped colour distances between the throat, chest, and breast regions of male Sicalis citrina.
</p>
</div>
</div>
<div id="converting-receptor-noise-corrected-distances-to-cartesian-coordinates" class="section level5" number="3.1.4.5.2">
<h5><span class="header-section-number">3.1.4.5.2</span> Converting receptor noise-corrected distances to Cartesian coordinates</h5>
<p>You can convert distances in JND back to Cartesian position coordinates with the function <code>jnd2xyz()</code>. Note that the actual position of these points in the XYZ space is arbitrary; therefore the function allows you to rotate the data so that, for example, the vector leading to the long-wavelength cone aligns with the X axis:</p>
<div class="sourceCode" id="cb133"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb133-1"><a href="analysing-data.html#cb133-1" aria-hidden="true" tabindex="-1"></a>fakedata1 <span class="ot">&lt;-</span> <span class="fu">vapply</span>(</span>
<span id="cb133-2"><a href="analysing-data.html#cb133-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">seq</span>(<span class="dv">100</span>, <span class="dv">500</span>, <span class="at">by =</span> <span class="dv">20</span>),</span>
<span id="cb133-3"><a href="analysing-data.html#cb133-3" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(x) <span class="fu">rowSums</span>(<span class="fu">cbind</span>(</span>
<span id="cb133-4"><a href="analysing-data.html#cb133-4" aria-hidden="true" tabindex="-1"></a>      <span class="fu">dnorm</span>(<span class="dv">300</span><span class="sc">:</span><span class="dv">700</span>, x, <span class="dv">30</span>),</span>
<span id="cb133-5"><a href="analysing-data.html#cb133-5" aria-hidden="true" tabindex="-1"></a>      <span class="fu">dnorm</span>(<span class="dv">300</span><span class="sc">:</span><span class="dv">700</span>, x <span class="sc">+</span> <span class="dv">400</span>, <span class="dv">30</span>)</span>
<span id="cb133-6"><a href="analysing-data.html#cb133-6" aria-hidden="true" tabindex="-1"></a>    )),</span>
<span id="cb133-7"><a href="analysing-data.html#cb133-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">numeric</span>(<span class="dv">401</span>)</span>
<span id="cb133-8"><a href="analysing-data.html#cb133-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb133-9"><a href="analysing-data.html#cb133-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-10"><a href="analysing-data.html#cb133-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Creating idealized specs with varying saturation</span></span>
<span id="cb133-11"><a href="analysing-data.html#cb133-11" aria-hidden="true" tabindex="-1"></a>fakedata2 <span class="ot">&lt;-</span> <span class="fu">vapply</span>(</span>
<span id="cb133-12"><a href="analysing-data.html#cb133-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="dv">500</span>, <span class="dv">300</span>, <span class="dv">150</span>, <span class="dv">105</span>, <span class="dv">75</span>, <span class="dv">55</span>, <span class="dv">40</span>, <span class="dv">30</span>),</span>
<span id="cb133-13"><a href="analysing-data.html#cb133-13" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(x) <span class="fu">dnorm</span>(<span class="dv">300</span><span class="sc">:</span><span class="dv">700</span>, <span class="dv">550</span>, x),</span>
<span id="cb133-14"><a href="analysing-data.html#cb133-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">numeric</span>(<span class="dv">401</span>)</span>
<span id="cb133-15"><a href="analysing-data.html#cb133-15" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb133-16"><a href="analysing-data.html#cb133-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-17"><a href="analysing-data.html#cb133-17" aria-hidden="true" tabindex="-1"></a>fakedata1 <span class="ot">&lt;-</span> <span class="fu">as.rspec</span>(<span class="fu">data.frame</span>(<span class="at">wl =</span> <span class="dv">300</span><span class="sc">:</span><span class="dv">700</span>, fakedata1))</span>
<span id="cb133-18"><a href="analysing-data.html#cb133-18" aria-hidden="true" tabindex="-1"></a>fakedata1 <span class="ot">&lt;-</span> <span class="fu">procspec</span>(fakedata1, <span class="st">&quot;max&quot;</span>)</span>
<span id="cb133-19"><a href="analysing-data.html#cb133-19" aria-hidden="true" tabindex="-1"></a>fakedata2 <span class="ot">&lt;-</span> <span class="fu">as.rspec</span>(<span class="fu">data.frame</span>(<span class="at">wl =</span> <span class="dv">300</span><span class="sc">:</span><span class="dv">700</span>, fakedata2))</span>
<span id="cb133-20"><a href="analysing-data.html#cb133-20" aria-hidden="true" tabindex="-1"></a>fakedata2 <span class="ot">&lt;-</span> <span class="fu">procspec</span>(fakedata2, <span class="st">&quot;sum&quot;</span>)</span>
<span id="cb133-21"><a href="analysing-data.html#cb133-21" aria-hidden="true" tabindex="-1"></a>fakedata2 <span class="ot">&lt;-</span> <span class="fu">procspec</span>(fakedata2, <span class="st">&quot;min&quot;</span>)</span>
<span id="cb133-22"><a href="analysing-data.html#cb133-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-23"><a href="analysing-data.html#cb133-23" aria-hidden="true" tabindex="-1"></a><span class="co"># Converting reflectance to percentage</span></span>
<span id="cb133-24"><a href="analysing-data.html#cb133-24" aria-hidden="true" tabindex="-1"></a>fakedata1[, <span class="sc">-</span><span class="dv">1</span>] <span class="ot">&lt;-</span> fakedata1[, <span class="sc">-</span><span class="dv">1</span>] <span class="sc">*</span> <span class="dv">100</span></span>
<span id="cb133-25"><a href="analysing-data.html#cb133-25" aria-hidden="true" tabindex="-1"></a>fakedata2[, <span class="sc">-</span><span class="dv">1</span>] <span class="ot">&lt;-</span> fakedata2[, <span class="sc">-</span><span class="dv">1</span>] <span class="sc">/</span> <span class="fu">max</span>(fakedata2[, <span class="sc">-</span><span class="dv">1</span>]) <span class="sc">*</span> <span class="dv">100</span></span>
<span id="cb133-26"><a href="analysing-data.html#cb133-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-27"><a href="analysing-data.html#cb133-27" aria-hidden="true" tabindex="-1"></a><span class="co"># Combining and converting to rspec</span></span>
<span id="cb133-28"><a href="analysing-data.html#cb133-28" aria-hidden="true" tabindex="-1"></a>fakedata.c <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">wl =</span> <span class="dv">300</span><span class="sc">:</span><span class="dv">700</span>, fakedata1[, <span class="sc">-</span><span class="dv">1</span>], fakedata2[, <span class="sc">-</span><span class="dv">1</span>])</span>
<span id="cb133-29"><a href="analysing-data.html#cb133-29" aria-hidden="true" tabindex="-1"></a>fakedata.c <span class="ot">&lt;-</span> <span class="fu">as.rspec</span>(fakedata.c)</span></code></pre></div>
<div class="sourceCode" id="cb134"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb134-1"><a href="analysing-data.html#cb134-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Visual model and colour distances</span></span>
<span id="cb134-2"><a href="analysing-data.html#cb134-2" aria-hidden="true" tabindex="-1"></a>fakedata.vm <span class="ot">&lt;-</span> <span class="fu">vismodel</span>(fakedata.c, <span class="at">relative =</span> <span class="cn">FALSE</span>, <span class="at">achromatic =</span> <span class="st">&#39;all&#39;</span>)</span>
<span id="cb134-3"><a href="analysing-data.html#cb134-3" aria-hidden="true" tabindex="-1"></a>fakedata.cd <span class="ot">&lt;-</span> <span class="fu">coldist</span>(fakedata.vm,</span>
<span id="cb134-4"><a href="analysing-data.html#cb134-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">noise =</span> <span class="st">&quot;neural&quot;</span>, <span class="at">n =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">4</span>),</span>
<span id="cb134-5"><a href="analysing-data.html#cb134-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">weber =</span> <span class="fl">0.1</span>, <span class="at">achromatic =</span> <span class="cn">TRUE</span></span>
<span id="cb134-6"><a href="analysing-data.html#cb134-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb134-7"><a href="analysing-data.html#cb134-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb134-8"><a href="analysing-data.html#cb134-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Converting to Cartesian coordinates</span></span>
<span id="cb134-9"><a href="analysing-data.html#cb134-9" aria-hidden="true" tabindex="-1"></a>fakedata.cc <span class="ot">&lt;-</span> <span class="fu">jnd2xyz</span>(fakedata.cd, <span class="at">ref1 =</span> <span class="st">&quot;l&quot;</span>, <span class="at">axis1 =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">0</span>), <span class="at">ref2 =</span> <span class="cn">NULL</span>)</span>
<span id="cb134-10"><a href="analysing-data.html#cb134-10" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(fakedata.cc)</span></code></pre></div>
<pre><code>##             x          y         z      lum
## X1 -26.165302   3.692783 -17.71718 1.430788
## X2 -11.212399  -2.085407 -27.97937 2.315951
## X3   3.076291  -7.441594 -37.76761 3.621349
## X4  16.523287 -12.475488 -46.33606 4.341465
## X5  25.812150 -17.848152 -40.71621 4.444027
## X6  31.912643 -23.368560 -24.79660 4.046863</code></pre>
<p>which you can then plot as well:</p>
<div class="sourceCode" id="cb136"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb136-1"><a href="analysing-data.html#cb136-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(fakedata.cc, <span class="at">theta =</span> <span class="dv">55</span>, <span class="at">phi =</span> <span class="dv">25</span>, <span class="at">col =</span> <span class="fu">spec2rgb</span>(fakedata.c))</span></code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-58"></span>
<img src="colsci-book_files/figure-html/unnamed-chunk-58-1.png" alt="Spectral data in a receptor noise-corrected colourspace" width="384" />
<p class="caption">
Figure 3.6: Spectral data in a receptor noise-corrected colourspace
</p>
</div>
<p>The axes in this colourspace are in JND units. For more information on these functions, see <code>?jnd2xyz</code>, <code>?jndrot</code> and <code>?jndplot</code>.</p>
</div>
</div>
<div id="colourspaces" class="section level4" number="3.1.4.6">
<h4><span class="header-section-number">3.1.4.6</span> Colourspaces</h4>
<p>Another general, flexible way in which to represent stimuli is via a colour space model. Under such models photon catches are expressed in relative values (so that the quantum catches of all receptors involved in chromatic discrimination sum to 1). The maximum stimulation of each cone <span class="math inline">\(n\)</span> is placed at the vertex of a <span class="math inline">\((n-1)\)</span>-dimensional polygon that encompasses all theoretical colours that can be perceived by that visual system. For the avian visual system comprised of four cones, for example, all colours can be placed somewhere in the volume of a tetrahedron, in which each of the four vertices represents the maximum stimulation of that particular cone type.</p>
<p>Though these models do not account for receptor noise (and thus do not allow an estimate of JNDs), they presents several advantages. First, they make for a very intuitive representation of colour points accounting for attributes of the colour vision of the signal receiver. Second, they allow for the calculation of several interesting variables that represent colour. For example, hue can be estimated from the angle of the point relative to the xy plane (blue-green-red) and the z axis (UV); saturation can be estimated as the distance of the point from the achromatic centre.</p>
<p><code>data(flowers)</code> is a new example dataset consisting of reflectance spectra from 36 Australian angiosperm species, which we’ll use to illustrate for the following colourspace modelling.</p>
<div class="sourceCode" id="cb137"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb137-1"><a href="analysing-data.html#cb137-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(flowers)</span>
<span id="cb137-2"><a href="analysing-data.html#cb137-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb137-3"><a href="analysing-data.html#cb137-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(flowers[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>])</span></code></pre></div>
<pre><code>##    wl Goodenia_heterophylla Goodenia_geniculata Goodenia_gracilis
## 1 300             1.7426387            1.918962         0.3638354
## 2 301             1.5724849            1.872966         0.3501921
## 3 302             1.4099808            1.828019         0.3366520
## 4 303             1.2547709            1.784152         0.3231911
## 5 304             1.1064997            1.741398         0.3097854
## 6 305             0.9648117            1.699792         0.2964107</code></pre>
<div id="di--tri--and-tetrachromatic-colourspaces" class="section level5" number="3.1.4.6.1">
<h5><span class="header-section-number">3.1.4.6.1</span> Di-, Tri-, and Tetrachromatic Colourspaces</h5>
<p><code>pavo</code> has extensive modelling and visualisation capabilities for generic di-, tri-, and tetra-chromatic spaces, uniting these approaches in a cohesive workflow. As with most colourspace models, we first estimate relative quantum catches with various assumptions by using the <code>vismodel()</code> function, before converting each set of values to a location in colourspace by using the <code>space</code> argument in <code>colspace()</code> (the function can also be set to try to detect the dimensionality of the colourspace automatically). For di- tri- and tetrachromatic spaces, <code>colspace()</code> calculates the coordinates of stimuli as:</p>
<p><strong>Dichromats:</strong>
<span class="math display">\[x = \frac{1}{\sqrt{2}}(Q_l - Q_s)\]</span></p>
<p><strong>Trichromats:</strong></p>
<p><span class="math display">\[
\begin{split}
x &amp;= \frac{1}{\sqrt{2}}(Q_l - Q_m) \\
y &amp;= \frac{\sqrt{2}}{\sqrt{3}}(Q_s - \frac{Q_l + Q_m}{2})
\end{split}
\]</span>
<strong>Tetrachromats:</strong></p>
<p><span class="math display">\[
\begin{split}
x &amp;= \frac{1}{\sqrt{2}}(Q_l - Q_m) \\
y &amp;= \frac{\sqrt{2}}{\sqrt{3}}(Q_s - \frac{Q_l + Q_m}{2})
z &amp;= \frac{\sqrt{3}}{2}(Q_u - \frac{Q_l + Q_m + Q_s}{3})
\end{split}
\]</span></p>
<p>Where <span class="math inline">\(Q_u\)</span>, <span class="math inline">\(Q_s\)</span>, <span class="math inline">\(Q_m\)</span>, and <span class="math inline">\(Q_l\)</span> refer to quantum catch estimates for UV-, short, medium-, and long-wavelength photoreceptors, respectively.</p>
<p>For a dichromatic example, we can model our floral reflectance data using the visual system of the domestic dog <em>Canis familiaris</em>, which has two cones with maximal sensitivity near 440 and 560 nm.</p>
<div class="sourceCode" id="cb139"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb139-1"><a href="analysing-data.html#cb139-1" aria-hidden="true" tabindex="-1"></a>vis.flowers <span class="ot">&lt;-</span> <span class="fu">vismodel</span>(flowers, <span class="at">visual =</span> <span class="st">&#39;canis&#39;</span>)</span>
<span id="cb139-2"><a href="analysing-data.html#cb139-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb139-3"><a href="analysing-data.html#cb139-3" aria-hidden="true" tabindex="-1"></a>di.flowers <span class="ot">&lt;-</span> <span class="fu">colspace</span>(vis.flowers, <span class="at">space =</span> <span class="st">&#39;di&#39;</span>)</span>
<span id="cb139-4"><a href="analysing-data.html#cb139-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb139-5"><a href="analysing-data.html#cb139-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(di.flowers)</span></code></pre></div>
<pre><code>##                                s         l           x      r.vec lum
## Goodenia_heterophylla 0.52954393 0.4704561 -0.04178143 0.04178143  NA
## Goodenia_geniculata   0.25430150 0.7456985  0.34747015 0.34747015  NA
## Goodenia_gracilis     0.01747832 0.9825217  0.68238870 0.68238870  NA
## Xyris_operculata      0.39433933 0.6056607  0.14942675 0.14942675  NA
## Eucalyptus_sp         0.40628552 0.5937145  0.13253229 0.13253229  NA
## Faradaya_splendida    0.23166580 0.7683342  0.37948187 0.37948187  NA</code></pre>
<p>The output contains values for the relative stimulation of short- and long-wavelength sensitive photoreceptors associated with each flower, along with its single coordinate in dichromatic space and its r.vector (distance from the origin). To visualise where these points lie, we can simply plot them on a segment.</p>
<div class="sourceCode" id="cb141"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb141-1"><a href="analysing-data.html#cb141-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(di.flowers, <span class="at">col =</span> <span class="fu">spec2rgb</span>(flowers))</span></code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-61"></span>
<img src="colsci-book_files/figure-html/unnamed-chunk-61-1.png" alt="Flowers in a dichromatic colourspace, as modelled according to a canid visual system." width="384" />
<p class="caption">
Figure 3.7: Flowers in a dichromatic colourspace, as modelled according to a canid visual system.
</p>
</div>
<p>For our trichromatic viewer we can use the honeybee <em>Apis mellifera</em>, one of the most significant and widespread pollinators. We’ll also transform our quantum catches according to Fechner’s law by specifying <code>qcatch = 'fi'</code>, and will model photoreceptor stimulation under bright conditions by scaling our illuminant with the <code>scale</code> argument.</p>
<div class="sourceCode" id="cb142"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb142-1"><a href="analysing-data.html#cb142-1" aria-hidden="true" tabindex="-1"></a>vis.flowers <span class="ot">&lt;-</span> <span class="fu">vismodel</span>(flowers, <span class="at">visual =</span> <span class="st">&#39;apis&#39;</span>, <span class="at">qcatch =</span> <span class="st">&#39;fi&#39;</span>, <span class="at">scale =</span> <span class="dv">10000</span>)</span>
<span id="cb142-2"><a href="analysing-data.html#cb142-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb142-3"><a href="analysing-data.html#cb142-3" aria-hidden="true" tabindex="-1"></a>tri.flowers <span class="ot">&lt;-</span> <span class="fu">colspace</span>(vis.flowers, <span class="at">space =</span> <span class="st">&#39;tri&#39;</span>)</span>
<span id="cb142-4"><a href="analysing-data.html#cb142-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb142-5"><a href="analysing-data.html#cb142-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(tri.flowers)</span></code></pre></div>
<pre><code>##                               s         m         l            x           y
## Goodenia_heterophylla 0.2882383 0.3587254 0.3530363 -0.004022828 -0.05522995
## Goodenia_geniculata   0.2475326 0.3549832 0.3974842  0.030052720 -0.10508396
## Goodenia_gracilis     0.1373117 0.3125304 0.5501579  0.168028001 -0.24007647
## Xyris_operculata      0.2395035 0.3729820 0.3875145  0.010275990 -0.11491764
## Eucalyptus_sp         0.2760091 0.3548096 0.3691813  0.010162377 -0.07020755
## Faradaya_splendida    0.2654951 0.3431266 0.3913783  0.034119147 -0.08308452
##                          h.theta      r.vec lum
## Goodenia_heterophylla -1.6435057 0.05537626  NA
## Goodenia_geniculata   -1.2922439 0.10929687  NA
## Goodenia_gracilis     -0.9601417 0.29303604  NA
## Xyris_operculata      -1.4816131 0.11537616  NA
## Eucalyptus_sp         -1.4270471 0.07093922  NA
## Faradaya_splendida    -1.1811377 0.08981733  NA</code></pre>
<p>As in the case of our dichromat, the output contains relative photoreceptor stimulations, coordinates in the Maxwell triangle, as well as the ‘hue angle’ <code>h.theta</code> and distance from the origin (<code>r.vec</code>).</p>
<div class="sourceCode" id="cb144"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb144-1"><a href="analysing-data.html#cb144-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tri.flowers, <span class="at">pch =</span> <span class="dv">21</span>, <span class="at">bg =</span> <span class="fu">spec2rgb</span>(flowers))</span></code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-63"></span>
<img src="colsci-book_files/figure-html/unnamed-chunk-63-1.png" alt="Floral reflectance in a Maxwell triangle, considering a honeybee visual system." width="384" />
<p class="caption">
Figure 3.8: Floral reflectance in a Maxwell triangle, considering a honeybee visual system.
</p>
</div>
<p>Finally, we’ll draw on the blue tit’s visual system to model our floral reflectance spectra in a tetrahedral space, again using log-transformed quantum catches and assuming bright viewing conditions.</p>
<div class="sourceCode" id="cb145"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb145-1"><a href="analysing-data.html#cb145-1" aria-hidden="true" tabindex="-1"></a>vis.flowers <span class="ot">&lt;-</span> <span class="fu">vismodel</span>(flowers, <span class="at">visual =</span> <span class="st">&quot;bluetit&quot;</span>, <span class="at">qcatch =</span> <span class="st">&quot;fi&quot;</span>, <span class="at">scale =</span> <span class="dv">10000</span>)</span>
<span id="cb145-2"><a href="analysing-data.html#cb145-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb145-3"><a href="analysing-data.html#cb145-3" aria-hidden="true" tabindex="-1"></a>tetra.flowers <span class="ot">&lt;-</span> <span class="fu">colspace</span>(vis.flowers, <span class="at">space =</span> <span class="st">&quot;tcs&quot;</span>)</span>
<span id="cb145-4"><a href="analysing-data.html#cb145-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb145-5"><a href="analysing-data.html#cb145-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(tetra.flowers)</span></code></pre></div>
<pre><code>##                               u         s         m         l         u.r
## Goodenia_heterophylla 0.2274310 0.2659669 0.2524170 0.2541851 -0.02256897
## Goodenia_geniculata   0.1625182 0.2721621 0.2828016 0.2825182 -0.08748184
## Goodenia_gracilis     0.0428552 0.2443091 0.3504414 0.3623943 -0.20714480
## Xyris_operculata      0.1766234 0.2709785 0.2642637 0.2881344 -0.07337659
## Eucalyptus_sp         0.2118481 0.2609468 0.2637023 0.2635028 -0.03815190
## Faradaya_splendida    0.1844183 0.2557660 0.2786393 0.2811764 -0.06558169
##                                s.r         m.r         l.r            x
## Goodenia_heterophylla  0.015966893 0.002417012 0.004185066 -0.007214866
## Goodenia_geniculata    0.022162064 0.032801599 0.032518179  0.006341799
## Goodenia_gracilis     -0.005690920 0.100441373 0.112394348  0.072312163
## Xyris_operculata       0.020978454 0.014263689 0.038134446  0.010505857
## Eucalyptus_sp          0.010946788 0.013702317 0.013502794  0.001565228
## Faradaya_splendida     0.005765955 0.028639328 0.031176408  0.015560661
##                                  y           z    h.theta     h.phi      r.vec
## Goodenia_heterophylla -0.005415708 -0.02256897 -2.4976873 -1.190529 0.02430520
## Goodenia_geniculata    0.003861848 -0.08748184  0.5469755 -1.486123 0.08779638
## Goodenia_gracilis      0.033297418 -0.20714480  0.4315247 -1.203879 0.22191605
## Xyris_operculata      -0.010813615 -0.07337659 -0.7998327 -1.368146 0.07490949
## Eucalyptus_sp          0.001044768 -0.03815190  0.5885699 -1.521510 0.03819828
## Faradaya_splendida     0.007189965 -0.06558169  0.4328380 -1.315140 0.06778487
##                           r.max r.achieved lum
## Goodenia_heterophylla 0.2692325 0.09027589  NA
## Goodenia_geniculata   0.2508989 0.34992737  NA
## Goodenia_gracilis     0.2678272 0.82857920  NA
## Xyris_operculata      0.2552227 0.29350636  NA
## Eucalyptus_sp         0.2503039 0.15260760  NA
## Faradaya_splendida    0.2583986 0.26232676  NA</code></pre>
<p>Tetrahedral data (via <code>colspace(space = 'tcs')</code>) may be plotted in a standard <em>static</em> tetrahedron by using <code>plot()</code>, or can be visualised as part of an <em>interactive</em> tetrahedron by using <code>tcsplot()</code> with the accessory functions <code>tcspoints()</code> and <code>tcsvol()</code> for adding points and convex hulls, respectively. As with other colourspace plots there are a number of associated graphical options, though the <code>theta</code> and <code>phi</code> arguments are particularly useful in this case, as they control the orientation (in degrees) of the tetrahedron in the xy and yz planes, respectively.</p>
<p>When plotting the tetrahedral colourspace, one can also force perspective by changing the size of points relative to their distance from the plane of observation, using the arguments <code>perspective = TRUE</code> and controlling the size range with the <code>range</code> argument. Several other options control the appearance of this plot, you can check these using <code>?plot.colspace</code> or <code>?tetraplot</code></p>
<div class="sourceCode" id="cb147"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb147-1"><a href="analysing-data.html#cb147-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tetra.flowers, <span class="at">pch =</span> <span class="dv">21</span>, <span class="at">bg =</span> <span class="fu">spec2rgb</span>(flowers), <span class="at">perspective =</span> <span class="cn">TRUE</span>, <span class="at">range =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">cex =</span> <span class="fl">0.5</span>)</span></code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-65"></span>
<img src="colsci-book_files/figure-html/unnamed-chunk-65-1.png" alt="Flowers in a tetrahedral colourspace modelled using the visual phenotype of the blue tit. Point size is used to force perspective" width="480" />
<p class="caption">
Figure 3.9: Flowers in a tetrahedral colourspace modelled using the visual phenotype of the blue tit. Point size is used to force perspective
</p>
</div>
<p>Two additional functions may help with tetrahedral colourspace plotting:</p>
<ul>
<li><code>axistetra()</code> function can be used to draw arrows showing the direction and magnitude of distortion of x, y and z in the tetrahedral plot.</li>
<li><code>legendtetra()</code> allows you to add a legend to a plot.</li>
</ul>
<div class="sourceCode" id="cb148"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb148-1"><a href="analysing-data.html#cb148-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tetra.flowers, <span class="at">pch =</span> <span class="dv">21</span>, <span class="at">bg =</span> <span class="fu">spec2rgb</span>(flowers))</span>
<span id="cb148-2"><a href="analysing-data.html#cb148-2" aria-hidden="true" tabindex="-1"></a><span class="fu">axistetra</span>(<span class="at">x =</span> <span class="dv">0</span>, <span class="at">y =</span> <span class="fl">1.8</span>)</span>
<span id="cb148-3"><a href="analysing-data.html#cb148-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tetra.flowers, <span class="at">theta =</span> <span class="dv">110</span>, <span class="at">phi =</span> <span class="dv">10</span>, <span class="at">pch =</span> <span class="dv">21</span>, <span class="at">bg =</span> <span class="fu">spec2rgb</span>(flowers))</span>
<span id="cb148-4"><a href="analysing-data.html#cb148-4" aria-hidden="true" tabindex="-1"></a><span class="fu">axistetra</span>(<span class="at">x =</span> <span class="dv">0</span>, <span class="at">y =</span> <span class="fl">1.8</span>)</span></code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-66"></span>
<img src="colsci-book_files/figure-html/unnamed-chunk-66-1.png" alt="Flowers in a tetrahedral colourspace, with varied orientations and perspectives, modelled using the visual phenotype of the blue tit." width="45%" /><img src="colsci-book_files/figure-html/unnamed-chunk-66-2.png" alt="Flowers in a tetrahedral colourspace, with varied orientations and perspectives, modelled using the visual phenotype of the blue tit." width="45%" />
<p class="caption">
Figure 3.10: Flowers in a tetrahedral colourspace, with varied orientations and perspectives, modelled using the visual phenotype of the blue tit.
</p>
</div>
<p>For tetrahedral models, another plotting option available is <code>projplot</code>, which projects colour points in the surface of a sphere encompassing the tetrahedron. This plot is particularly useful to see differences in hue.</p>
<div class="sourceCode" id="cb149"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb149-1"><a href="analysing-data.html#cb149-1" aria-hidden="true" tabindex="-1"></a><span class="fu">projplot</span>(tetra.flowers, <span class="at">pch =</span> <span class="dv">20</span>, <span class="at">col =</span> <span class="fu">spec2rgb</span>(flowers))</span></code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-67"></span>
<img src="colsci-book_files/figure-html/unnamed-chunk-67-1.png" alt="Projection plot from a tetrahedral colour space." width="480" />
<p class="caption">
Figure 3.11: Projection plot from a tetrahedral colour space.
</p>
</div>
<p>Finally, a useful function for tri- and tetrahedral models is <code>voloverlap()</code>, which calculates the overlap in tri- or tetrahedral colour volume between two sets of points. This can be useful to explore whether different species occupy similar (overlapping) or different (non- overlapping) “sensory niches,” or to test for mimetism, dichromatism, etc. <span class="citation">(<a href="#ref-Stoddard2011" role="doc-biblioref">Stoddard and Stevens 2011</a>)</span>. To show this function, we will use the <code>sicalis</code> dataset, which includes measurements from the crown (C), throat (T) and breast (B) of seven stripe-tailed yellow finches (<em>Sicalis citrina</em>).</p>
<div class="sourceCode" id="cb150"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb150-1"><a href="analysing-data.html#cb150-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(sicalis)</span></code></pre></div>
<p>We will use this dataset to test for the overlap between the volume determined by the measurements of those body parts from multiple individuals in the tetrahedral colourspace (note the option <code>plot</code> for plotting of the volumes):</p>
<div class="sourceCode" id="cb151"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb151-1"><a href="analysing-data.html#cb151-1" aria-hidden="true" tabindex="-1"></a>tcs.sicalis.C <span class="ot">&lt;-</span> <span class="fu">subset</span>(<span class="fu">colspace</span>(<span class="fu">vismodel</span>(sicalis)), <span class="st">&quot;C&quot;</span>)</span>
<span id="cb151-2"><a href="analysing-data.html#cb151-2" aria-hidden="true" tabindex="-1"></a>tcs.sicalis.T <span class="ot">&lt;-</span> <span class="fu">subset</span>(<span class="fu">colspace</span>(<span class="fu">vismodel</span>(sicalis)), <span class="st">&quot;T&quot;</span>)</span>
<span id="cb151-3"><a href="analysing-data.html#cb151-3" aria-hidden="true" tabindex="-1"></a>tcs.sicalis.B <span class="ot">&lt;-</span> <span class="fu">subset</span>(<span class="fu">colspace</span>(<span class="fu">vismodel</span>(sicalis)), <span class="st">&quot;B&quot;</span>)</span>
<span id="cb151-4"><a href="analysing-data.html#cb151-4" aria-hidden="true" tabindex="-1"></a><span class="fu">voloverlap</span>(tcs.sicalis.T, tcs.sicalis.B, <span class="at">plot =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>##           vol1         vol2   overlapvol vsmallest      vboth
## 1 5.183721e-06 6.281511e-06 6.904074e-07 0.1331876 0.06407598</code></pre>
<div class="sourceCode" id="cb153"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb153-1"><a href="analysing-data.html#cb153-1" aria-hidden="true" tabindex="-1"></a><span class="fu">voloverlap</span>(tcs.sicalis.T, tcs.sicalis.C, <span class="at">plot =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>##           vol1         vol2 overlapvol vsmallest vboth
## 1 5.183721e-06 4.739152e-06          0         0     0</code></pre>
<p><img src="colsci-book_files/figure-html/unnamed-chunk-69-1.png" width="40%" /><img src="colsci-book_files/figure-html/unnamed-chunk-69-2.png" width="40%" /></p>
<p>The function <code>voloverlap()</code> gives the volume (<span class="math inline">\(V\)</span>) of the convex hull delimited by the overlap between the two original volumes, and two proportions are calculated from that: <span class="math display">\[\text{vsmallest} = V_{overlap} / \min(V_A, V_B)\]</span> and <span class="math display">\[\text{vboth} = V_{overlap} / (V_A + V_B)\]</span>. Thus, if one of the volumes is entirely contained in the other, <code>vsmallest</code> will equal 1.</p>
<p>So we can clearly see that there is overlap between the throat and breast colours (of about 6%), but not between the throat and the crown colours (Figures above).</p>
<p><strong>Summary variables for groups of points</strong></p>
<p>Another advantage of colourspace models is that they allow for the calculation of useful summary statistics of groups of points, such as the centroid of the points, the total volume occupied, the mean and variance of hue span and the mean saturation. In <code>pavo</code>, the result of a <code>colspace()</code> call is an object of class <code>colspace</code>, and thus these summary statistics can be calculated simply by calling <code>summary</code>. Note that the <code>summary</code> call can also take a <code>by</code> vector of group identities, so that the variables are calculated for each group separately:</p>
<div class="sourceCode" id="cb155"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb155-1"><a href="analysing-data.html#cb155-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(tetra.flowers)</span></code></pre></div>
<pre><code>## Colorspace &amp; visual model options:
##  * Colorspace: tcs 
##  * Quantal catch: fi 
##  * Visual system, chromatic: bluetit 
##  * Visual system, achromatic: none 
##  * Illuminant: ideal, scale = 10000 (von Kries colour correction not applied) 
##  * Background: ideal 
##  * Relative: TRUE 
##  * Max possible chromatic volume: NA</code></pre>
<pre><code>## &#39;avalue&#39; automatically set to 1.9861e-01</code></pre>
<pre><code>##            centroid.u centroid.s centroid.m centroid.l        c.vol   rel.c.vol
## all.points  0.1948478  0.2554421  0.2704411   0.279269 0.0002510915 0.001159742
##             colspan.m   colspan.v huedisp.m huedisp.v   mean.ra    max.ra
## all.points 0.05706825 0.001682224 0.6413439 0.3157949 0.2428508 0.8285792
##                   a.vol
## all.points 0.0001723672</code></pre>
<p>It is important to highlight that oftentimes, the properties of the visual system will make it impossible to reach the entirety of the Maxwell triangle space (or tetrahedron for tetrachromatic viewers). You can view the maximum gamut of a given viewer by switching the <code>gamut</code> argument to <code>TRUE</code> in <code>plot.colspace()</code>:</p>
<div class="sourceCode" id="cb159"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb159-1"><a href="analysing-data.html#cb159-1" aria-hidden="true" tabindex="-1"></a>vis.flowers <span class="ot">&lt;-</span> <span class="fu">vismodel</span>(flowers, <span class="at">visual =</span> <span class="st">&quot;ctenophorus&quot;</span>)</span>
<span id="cb159-2"><a href="analysing-data.html#cb159-2" aria-hidden="true" tabindex="-1"></a>tri.flowers <span class="ot">&lt;-</span> <span class="fu">colspace</span>(vis.flowers)</span>
<span id="cb159-3"><a href="analysing-data.html#cb159-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tri.flowers, <span class="at">pch =</span> <span class="dv">21</span>, <span class="at">bg =</span> <span class="fu">spec2rgb</span>(flowers), <span class="at">gamut =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="colsci-book_files/figure-html/unnamed-chunk-71-1.png" width="672" /></p>
<p>This information is also available in the summary, in the “Max possible chromatic volume” line:</p>
<div class="sourceCode" id="cb160"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb160-1"><a href="analysing-data.html#cb160-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(tri.flowers)</span></code></pre></div>
<pre><code>## Colorspace &amp; visual model options:
##  * Colorspace: trispace 
##  * Quantal catch: Qi 
##  * Visual system, chromatic: ctenophorus 
##  * Visual system, achromatic: none 
##  * Illuminant: ideal, scale = 1 (von Kries colour correction not applied) 
##  * Background: ideal 
##  * Relative: TRUE 
##  * Max possible chromatic volume: 0.4729301</code></pre>
<pre><code>##        s                 m                l                x           
##  Min.   :0.02084   Min.   :0.2318   Min.   :0.2066   Min.   :-0.09877  
##  1st Qu.:0.18617   1st Qu.:0.3005   1st Qu.:0.3806   1st Qu.: 0.02242  
##  Median :0.23206   Median :0.3452   Median :0.4180   Median : 0.05899  
##  Mean   :0.23206   Mean   :0.3225   Mean   :0.4454   Mean   : 0.08689  
##  3rd Qu.:0.29330   3rd Qu.:0.3513   3rd Qu.:0.5039   3rd Qu.: 0.12564  
##  Max.   :0.44720   Max.   :0.3660   Max.   :0.7474   Max.   : 0.36460  
##        y               h.theta            r.vec           lum         
##  Min.   :-0.38273   Min.   :-1.9491   Min.   :0.02264   Mode:logical  
##  1st Qu.:-0.18023   1st Qu.:-1.2269   1st Qu.:0.08330   NA&#39;s:36       
##  Median :-0.12404   Median :-0.9590   Median :0.13917                 
##  Mean   :-0.12404   Mean   :-0.5767   Mean   :0.17858                 
##  3rd Qu.:-0.04903   3rd Qu.:-0.7747   3rd Qu.:0.24002                 
##  Max.   : 0.13946   Max.   : 2.6273   Max.   :0.52860</code></pre>
</div>
<div id="the-colour-hexagon" class="section level5" number="3.1.4.6.2">
<h5><span class="header-section-number">3.1.4.6.2</span> The Colour Hexagon</h5>
<p>The hexagon colour space of <span class="citation"><a href="#ref-Chittka1992" role="doc-biblioref">Chittka</a> (<a href="#ref-Chittka1992" role="doc-biblioref">1992</a>)</span> is a generalised colour-opponent model of hymenopteran vision that has found extremely broad use, particularly in studies of bee-flower interactions. It’s also often broadly applied across hymenopteran species, because the photopigments underlying trichromatic vision in Hymenoptera appear to be quite conserved <span class="citation">(<a href="#ref-Briscoe2001" role="doc-biblioref">Briscoe and Chittka 2001</a>)</span>. What’s particularly useful is that colour distances within the hexagon have been extensively validated against behaviour, and thus offer a relatively reliable measure of perceptual distance.</p>
<p>In the hexagon, photoreceptor quantum catches are typically hyperbolically transformed (and <code>pavo</code> will return a warning if the transform is not selected), and vonkries correction is often used used to model photoreceptor adaptation to a vegetation background. This can all now be specified in <code>vismodel()</code>. including the optional use of a ‘green’ vegetation background. Note that although this is a colourspace model, we specific <code>relative = FALSE</code> to return unnormalised quantum catches, as required for the model.</p>
<div class="sourceCode" id="cb163"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb163-1"><a href="analysing-data.html#cb163-1" aria-hidden="true" tabindex="-1"></a>vis.flowers <span class="ot">&lt;-</span> <span class="fu">vismodel</span>(flowers, <span class="at">visual =</span> <span class="st">&#39;apis&#39;</span>, <span class="at">qcatch =</span> <span class="st">&#39;Ei&#39;</span>, <span class="at">relative =</span> <span class="cn">FALSE</span>, <span class="at">vonkries =</span> <span class="cn">TRUE</span>, <span class="at">achromatic =</span> <span class="st">&#39;l&#39;</span>, <span class="at">bkg =</span> <span class="st">&#39;green&#39;</span>)</span></code></pre></div>
<p>We can then apply the hexagon model in <code>colspace</code>, which will convert our photoreceptor ‘excitation values’ to coordinates in the hexagon according to:</p>
<p><span class="math display">\[
\begin{split}
x &amp;= \frac{\sqrt{3}}{2(E_g + E_{uv})}\\
y &amp;= E_b - 0.5(E_{uv} + E_g)
\end{split}
\]</span></p>
<div class="sourceCode" id="cb164"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb164-1"><a href="analysing-data.html#cb164-1" aria-hidden="true" tabindex="-1"></a>hex.flowers <span class="ot">&lt;-</span> <span class="fu">colspace</span>(vis.flowers, <span class="at">space =</span> <span class="st">&#39;hexagon&#39;</span>)</span>
<span id="cb164-2"><a href="analysing-data.html#cb164-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb164-3"><a href="analysing-data.html#cb164-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(hex.flowers)</span></code></pre></div>
<pre><code>##                                s         m         l         x          y
## Goodenia_heterophylla 0.56572097 0.8237141 0.7053057 0.1208839  0.1882007
## Goodenia_geniculata   0.35761236 0.8176153 0.8670134 0.4411542  0.2053024
## Goodenia_gracilis     0.01888788 0.1622766 0.7810589 0.6600594 -0.2376968
## Xyris_operculata      0.21080752 0.7345122 0.6796464 0.4060263  0.2892853
## Eucalyptus_sp         0.55622758 0.8515289 0.8208038 0.2291297  0.1630132
## Faradaya_splendida    0.45855056 0.7828905 0.8565895 0.3447118  0.1253205
##                         h.theta     r.vec sec.fine sec.coarse       lum
## Goodenia_heterophylla  32.71320 0.2236793       30  bluegreen 0.1680153
## Goodenia_geniculata    65.04388 0.4865862       60  bluegreen 0.3548825
## Goodenia_gracilis     109.80467 0.7015541      100      green 0.2313678
## Xyris_operculata       54.53092 0.4985412       50  bluegreen 0.1518319
## Eucalyptus_sp          54.57024 0.2812005       50  bluegreen 0.2787543
## Faradaya_splendida     70.02119 0.3667853       70  bluegreen 0.3351008</code></pre>
<p>Again, the output includes the photoreceptor excitation values for short- medium- and long-wave sensitive photoreceptors, x and y coordinates, and measures of hue and saturation for each stimulus. The hexagon model also outputs two additional measures of of subjective ‘bee-hue’; <code>sec.fine</code> and <code>sec.coarse</code>. <code>sec.fine</code> describes the location of stimuli within one of 36 ‘hue sectors’ that are specified by radially dissecting the hexagon in 10-degree increments. <code>sec.coarse</code> follows a similar principle, though here the hexagon is divided into only five ‘bee-hue’ sectors: UV, UV-blue, blue, blue-green, green, and UV-green. These can easily be visualised by specifying <code>sectors = 'coarse</code> or <code>sectors = 'fine'</code> in a call to <code>plot</code> after modelling.</p>
<div class="sourceCode" id="cb166"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb166-1"><a href="analysing-data.html#cb166-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hex.flowers, <span class="at">sectors =</span> <span class="st">&#39;coarse&#39;</span>, <span class="at">pch =</span> <span class="dv">21</span>, <span class="at">bg =</span> <span class="fu">spec2rgb</span>(flowers))</span></code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-75"></span>
<img src="colsci-book_files/figure-html/unnamed-chunk-75-1.png" alt="Flowers as modelled in the hymenopteran colour hexagon of Chittka (1992), overlain with coarse bee-hue sectors." width="480" />
<p class="caption">
Figure 3.12: Flowers as modelled in the hymenopteran colour hexagon of Chittka (1992), overlain with coarse bee-hue sectors.
</p>
</div>
</div>
<div id="the-colour-opponent-coding-coc-space" class="section level5" number="3.1.4.6.3">
<h5><span class="header-section-number">3.1.4.6.3</span> The Colour Opponent Coding (COC) Space</h5>
<p>The colour opponent coding (coc) space is an earlier hymenopteran visual model <span class="citation">(<a href="#ref-Backhaus1991" role="doc-biblioref">Backhaus 1991</a>)</span> which, although now seldom used in favour of the hexagon, may prove useful for comparative work. While the initial estimation of photoreceptor excitation is similar to that in the hexagon, the coc subsequently specifies <code>A</code> and <code>B</code> coordinates based on empirically-derived weights for the output from each photoreceptor:</p>
<p><span class="math display">\[
\begin{split}
A &amp;= -9.86E_u + 7.70E_b + 2.16E_g\\
B &amp;= -5.17E_u + 20.25E_b - 15.08E_g
\end{split}
\]</span></p>
<p>Where <span class="math inline">\(E_i\)</span> is the excitation value (hyperbolic-transformed quantum catch) in photoreceptor <span class="math inline">\(i\)</span>.</p>
<div class="sourceCode" id="cb167"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb167-1"><a href="analysing-data.html#cb167-1" aria-hidden="true" tabindex="-1"></a>vis.flowers <span class="ot">&lt;-</span> <span class="fu">vismodel</span>(flowers, <span class="at">visual =</span> <span class="st">&quot;apis&quot;</span>, <span class="at">qcatch =</span> <span class="st">&quot;Ei&quot;</span>, <span class="at">relative =</span> <span class="cn">FALSE</span>, <span class="at">vonkries =</span> <span class="cn">TRUE</span>, <span class="at">bkg =</span> <span class="st">&quot;green&quot;</span>)</span>
<span id="cb167-2"><a href="analysing-data.html#cb167-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb167-3"><a href="analysing-data.html#cb167-3" aria-hidden="true" tabindex="-1"></a>coc.flowers <span class="ot">&lt;-</span> <span class="fu">colspace</span>(vis.flowers, <span class="at">space =</span> <span class="st">&quot;coc&quot;</span>)</span>
<span id="cb167-4"><a href="analysing-data.html#cb167-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb167-5"><a href="analysing-data.html#cb167-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(coc.flowers)</span></code></pre></div>
<pre><code>##                                s         m         l        x          y
## Goodenia_heterophylla 0.56572097 0.8237141 0.7053057 2.288050  3.1194224
## Goodenia_geniculata   0.35761236 0.8176153 0.8670134 4.642329  1.6332926
## Goodenia_gracilis     0.01888788 0.1622766 0.7810589 2.750382 -8.5899171
## Xyris_operculata      0.21080752 0.7345122 0.6796464 5.045218  3.5349309
## Eucalyptus_sp         0.55622758 0.8515289 0.8208038 2.845304  1.9900420
## Faradaya_splendida    0.45855056 0.7828905 0.8565895 3.357182  0.5654568
##                           r.vec lum
## Goodenia_heterophylla  5.407473  NA
## Goodenia_geniculata    6.275622  NA
## Goodenia_gracilis     11.340299  NA
## Xyris_operculata       8.580149  NA
## Eucalyptus_sp          4.835346  NA
## Faradaya_splendida     3.922638  NA</code></pre>
<p>The A and B coordinates are designated x and y in the output of coc for consistency, and while the model includes a measure of saturation in <code>r.vec</code>, it contains no associated measure of hue.</p>
<div class="sourceCode" id="cb169"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb169-1"><a href="analysing-data.html#cb169-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(coc.flowers, <span class="at">pch =</span> <span class="dv">21</span>, <span class="at">bg =</span> <span class="fu">spec2rgb</span>(flowers), <span class="at">yaxt =</span> <span class="st">&quot;n&quot;</span>)</span></code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-77"></span>
<img src="colsci-book_files/figure-html/unnamed-chunk-77-1.png" alt="Flowers in the colour-opponent-coding space of Backhaus (1991), as modelling according to the honeybee." width="480" />
<p class="caption">
Figure 3.13: Flowers in the colour-opponent-coding space of Backhaus (1991), as modelling according to the honeybee.
</p>
</div>
</div>
<div id="cie-spaces" class="section level5" number="3.1.4.6.4">
<h5><span class="header-section-number">3.1.4.6.4</span> CIE Spaces</h5>
<p>The CIE (International Commission on Illumination) colourspaces are a suite of models of human colour vision and perception. <code>pavo 1.0</code> now includes two of the most commonly used: the foundational 1931 CIE XYZ space, and the more modern, perceptually calibrated CIE LAB space and its cylindrical CIE LCh transformation. Tristimulus values in XYZ space are calculated as:</p>
<p><span class="math display">\[
\begin{split}
X &amp;= k\int_{300}^{700}{R(\lambda)I(\lambda)\bar{x}(\lambda)\,d\lambda}\\
Y &amp;= k\int_{300}^{700}{R(\lambda)I(\lambda)\bar{y}(\lambda)\,d\lambda}\\
Z &amp;= k\int_{300}^{700}{R(\lambda)I(\lambda)\bar{z}(\lambda)\,d\lambda}
\end{split}
\]</span></p>
<p>where <em>x</em>, <em>y</em>, and <em>z</em> are the trichromatic colour matching functions for a ‘standard colourimetric viewer.’ These functions are designed to describe an average human’s chromatic response within a specified viewing arc in the fovea (to account for the uneven distribution of cones across eye). <code>pavo</code> includes both the CIE 2-degree and the modern 10-degree standard observer, which can be selected in the <code>visual</code> option in the <code>vismodel</code> function. In these equations, k is the normalising factor</p>
<p><span class="math display">\[k = \frac{100}{\int_{300}^{700}{I(\lambda)\bar{y}(\lambda)\,d\lambda}}\]</span></p>
<p>and the chromaticity coordinates of stimuli are calculated as</p>
<p><span class="math display">\[
\begin{split}
x &amp;= \frac{X}{X + Y + Z}\\
y &amp;= \frac{Y}{X + Y + Z}\\
z &amp;= \frac{Z}{X + Y + Z} = 1 - x - y
\end{split}
\]</span></p>
<p>For modelling in both XYZ and LAB spaces, here we’ll use the CIE 10-degree standard observer, and assume a <code>D65</code> ‘standard daylight’ illuminant. Again, although they are colourspace models, we need to set <code>relative = FALSE</code> to return raw quantum catch estimates, <code>vonkries = TRUE</code> to account for the required normalising factor (as above), and <code>achromatic = 'none'</code> since there is no applicable way to estimate luminance in the CIE models. As with all models that have particular requirements, <code>vismodel()</code> will output warnings and/or errors if unusual or non-standard arguments are specified.</p>
<div class="sourceCode" id="cb170"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb170-1"><a href="analysing-data.html#cb170-1" aria-hidden="true" tabindex="-1"></a>vis.flowers <span class="ot">&lt;-</span> <span class="fu">vismodel</span>(flowers, <span class="at">visual =</span> <span class="st">&#39;cie10&#39;</span>, <span class="at">illum =</span> <span class="st">&#39;D65&#39;</span>, <span class="at">vonkries =</span> <span class="cn">TRUE</span>, <span class="at">relative =</span> <span class="cn">FALSE</span>, <span class="at">achromatic =</span> <span class="st">&#39;none&#39;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb171"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb171-1"><a href="analysing-data.html#cb171-1" aria-hidden="true" tabindex="-1"></a>ciexyz.flowers <span class="ot">&lt;-</span> <span class="fu">colspace</span>(vis.flowers, <span class="at">space =</span> <span class="st">&#39;ciexyz&#39;</span>)</span>
<span id="cb171-2"><a href="analysing-data.html#cb171-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ciexyz.flowers)</span></code></pre></div>
<pre><code>##                               X         Y          Z         x         y
## Goodenia_heterophylla 0.2095846 0.2074402 0.30060475 0.2920513 0.2890630
## Goodenia_geniculata   0.6415362 0.6751982 0.42423209 0.3684943 0.3878295
## Goodenia_gracilis     0.5027874 0.4538738 0.01742707 0.5161620 0.4659473
## Xyris_operculata      0.2847275 0.2325247 0.22131809 0.3855117 0.3148309
## Eucalyptus_sp         0.4421178 0.4469747 0.40173521 0.3425072 0.3462699
## Faradaya_splendida    0.6545883 0.6532237 0.28855528 0.4100487 0.4091939
##                                z
## Goodenia_heterophylla 0.41888569
## Goodenia_geniculata   0.24367620
## Goodenia_gracilis     0.01789065
## Xyris_operculata      0.29965743
## Eucalyptus_sp         0.31122295
## Faradaya_splendida    0.18075745</code></pre>
<p>The output is simply the tristimulus values and chromaticity coordinates of stimuli, and we can visualise our results (along with a line connecting monochromatic loci, by default) by calling</p>
<div class="sourceCode" id="cb173"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb173-1"><a href="analysing-data.html#cb173-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(ciexyz.flowers, <span class="at">pch =</span> <span class="dv">21</span>, <span class="at">bg =</span> <span class="fu">spec2rgb</span>(flowers))</span></code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-80"></span>
<img src="colsci-book_files/figure-html/unnamed-chunk-80-1.png" alt="Floral reflectance in the CIEXYZ human visual model. Note that this space is not perceptually calibrated, so we cannot make inferences about the similarity or differences of colours based on their relative location." width="480" />
<p class="caption">
Figure 3.14: Floral reflectance in the CIEXYZ human visual model. Note that this space is not perceptually calibrated, so we cannot make inferences about the similarity or differences of colours based on their relative location.
</p>
</div>
<p>The <code>Lab</code> space is a more recent development, and is a colour-opponent model that attempts to mimic the nonlinear responses of the human eye. The <code>Lab</code> space is also calibrated (unlike the <code>XYZ</code> space) such that Euclidean distances between points represent relative perceptual distances. This means that two stimuli that are farther apart in <code>Lab</code> space should be <em>perceived</em> as more different than two closer points. As the name suggests, the dimensions of the <code>Lab</code> space are <em>L</em>ightness (i.e. subjective brightness), along with two colour-opponent dimensions designated <em>a</em> and <em>b</em>. The <code>colspace()</code> function, when <code>space = cielab</code>, simply converts points from the <code>XYZ</code> model according to:</p>
$$
<span class="math display">\[\begin{split}

L &amp;=\begin{cases}
116\left(\frac{Y}{Y_n}\right)^\frac{1}{3} &amp; \text{if } \frac{Y}{Y_n} &gt; 0.008856 \\
903.3\left(\frac{Y}{Y_n}\right) &amp; \text{if } \frac{Y}{Y_n} \leq 0.008856
\end{cases}\\

a &amp;= 500\left(f\left(\frac{X}{X_n}\right) - f\left(\frac{Y}{Y_n}\right)\right)\\

b &amp;= 500\left(f\left(\frac{Y}{Y_n}\right) - f\left(\frac{Z}{Z_n}\right)\right)

\end{split}\]</span>
<p>$$</p>
<p>where</p>
<p><span class="math display">\[
f(x)=\begin{cases}
x^\frac{1}{3} &amp; \text{if } x &gt; 0.008856 \\
7.787x + \frac{4}{29} &amp; \text{if } x \leq 0.008856
\end{cases}
\]</span></p>
<p>Here, <span class="math inline">\(X_n, Y_n, Z_n\)</span> are neutral point values to model visual adaptation, calculated as:</p>
<p><span class="math display">\[
\begin{split}
X_n &amp;= \int_{300}^{700}{R_n(\lambda)I(\lambda)\bar{x}(\lambda)\,d\lambda}\\
Y_n &amp;= \int_{300}^{700}{R_n(\lambda)I(\lambda)\bar{y}(\lambda)\,d\lambda}\\
Z_n &amp;= \int_{300}^{700}{R_n(\lambda)I(\lambda)\bar{z}(\lambda)\,d\lambda}
\end{split}
\]</span></p>
<p>when <span class="math inline">\(R_n(\lambda)\)</span> is a perfect diffuse reflector (i.e. 1).</p>
<div class="sourceCode" id="cb174"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb174-1"><a href="analysing-data.html#cb174-1" aria-hidden="true" tabindex="-1"></a>cielab.flowers <span class="ot">&lt;-</span> <span class="fu">colspace</span>(vis.flowers, <span class="at">space =</span> <span class="st">&#39;cielab&#39;</span>)</span>
<span id="cb174-2"><a href="analysing-data.html#cb174-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(cielab.flowers)</span></code></pre></div>
<pre><code>##                               X         Y          Z        L         a
## Goodenia_heterophylla 0.2095846 0.2074402 0.30060475 16.75107  3.110872
## Goodenia_geniculata   0.6415362 0.6751982 0.42423209 54.52316 -4.478971
## Goodenia_gracilis     0.5027874 0.4538738 0.01742707 36.65092 22.697331
## Xyris_operculata      0.2847275 0.2325247 0.22131809 18.77668 21.382120
## Eucalyptus_sp         0.4421178 0.4469747 0.40173521 36.09381  3.297259
## Faradaya_splendida    0.6545883 0.6532237 0.28855528 52.74869  7.859790
##                                b
## Goodenia_heterophylla -18.616834
## Goodenia_geniculata    26.973109
## Goodenia_gracilis      60.437370
## Xyris_operculata       -2.596506
## Eucalyptus_sp          -1.246816
## Faradaya_splendida     45.351679</code></pre>
<p>Our output now contains the tristimulus <code>XYZ</code> values, as well as their <code>Lab</code> counterparts, which are coordinates in the <code>Lab</code> space. These can also be visualised in three-dimensional <code>Lab</code> space by calling <code>plot</code>:</p>
<div class="sourceCode" id="cb176"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb176-1"><a href="analysing-data.html#cb176-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(cielab.flowers, <span class="at">pch =</span> <span class="dv">21</span>, <span class="at">bg =</span> <span class="fu">spec2rgb</span>(flowers))</span></code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-82"></span>
<img src="colsci-book_files/figure-html/unnamed-chunk-82-1.png" alt="Floral reflectance spectra represented in the CIELab model of human colour sensation." width="480" />
<p class="caption">
Figure 3.15: Floral reflectance spectra represented in the CIELab model of human colour sensation.
</p>
</div>
</div>
<div id="categorical-fly-colourspace" class="section level5" number="3.1.4.6.5">
<h5><span class="header-section-number">3.1.4.6.5</span> Categorical Fly Colourspace</h5>
<p>The categorical colour vision model of <span class="citation"><a href="#ref-Troje1993" role="doc-biblioref">Troje</a> (<a href="#ref-Troje1993" role="doc-biblioref">1993</a>)</span> is a model of dipteran vision, based on behavioural data from the blowfly <em>Lucilia</em> sp. It assumes the involvement of all four dipteran photoreceptor classes (R7 &amp; R8 ‘pale’ and ‘yellow’ subtypes), and further posits that colour vision is based on two specific opponent mechanisms (R7p - R8p, and R7y - R8y). The model assumes that all colours are perceptually grouped into one of four colour categories, and that flies are unable to distinguish between colours that fall within the same category.</p>
<p>We’ll use the visual systems of the muscoid fly <em>Musca domestica</em>, and will begin by estimating linear (i.e. untransformed) quantum catches for each of the four photoreceptors.</p>
<div class="sourceCode" id="cb177"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb177-1"><a href="analysing-data.html#cb177-1" aria-hidden="true" tabindex="-1"></a>vis.flowers <span class="ot">&lt;-</span> <span class="fu">vismodel</span>(flowers, <span class="at">qcatch =</span> <span class="st">&#39;Qi&#39;</span>, <span class="at">visual =</span> <span class="st">&#39;musca&#39;</span>, <span class="at">achromatic =</span> <span class="st">&#39;none&#39;</span>, <span class="at">relative =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>Our call to <code>colspace()</code> will then simply estimate the location of stimuli in the categorical space as the difference in relative stimulation between ‘pale’ (R7p - R8p) and ‘yellow’ (R7y - R8y) photoreceptor pairs:</p>
<p><span class="math display">\[
\begin{split}
x &amp;= R7_p - R8_p\\
y &amp;= R7_y - R8_y
\end{split}
\]</span></p>
<div class="sourceCode" id="cb178"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb178-1"><a href="analysing-data.html#cb178-1" aria-hidden="true" tabindex="-1"></a>cat.flowers <span class="ot">&lt;-</span> <span class="fu">colspace</span>(vis.flowers, <span class="at">space =</span> <span class="st">&#39;categorical&#39;</span>)</span>
<span id="cb178-2"><a href="analysing-data.html#cb178-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-3"><a href="analysing-data.html#cb178-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(cat.flowers)</span></code></pre></div>
<pre><code>##                              R7p        R7y       R8p       R8y          x
## Goodenia_heterophylla 0.05387616 0.18710662 0.4335986 0.3254187 -0.3797224
## Goodenia_geniculata   0.02080098 0.08276030 0.3737592 0.5226795 -0.3529583
## Goodenia_gracilis     0.01153130 0.02501417 0.1420049 0.8214496 -0.1304736
## Xyris_operculata      0.01270353 0.12523419 0.4488377 0.4132246 -0.4361342
## Eucalyptus_sp         0.03461949 0.14207848 0.3977208 0.4255812 -0.3631013
## Faradaya_splendida    0.03599953 0.09188560 0.3129195 0.5591953 -0.2769200
##                                y category     r.vec   h.theta lum
## Goodenia_heterophylla -0.1383120     p-y- 0.4041279 -2.792284  NA
## Goodenia_geniculata   -0.4399192     p-y- 0.5640110 -2.246954  NA
## Goodenia_gracilis     -0.7964355     p-y- 0.8070519 -1.733176  NA
## Xyris_operculata      -0.2879904     p-y- 0.5226390 -2.557993  NA
## Eucalyptus_sp         -0.2835027     p-y- 0.4606695 -2.478681  NA
## Faradaya_splendida    -0.4673097     p-y- 0.5431971 -2.105745  NA</code></pre>
<p>And it is simply the signs of these differences that define the four possible fly-colour categories (p+y+, p-y+, p+y-, p-y-), which we can see in the associated plot.</p>
<div class="sourceCode" id="cb180"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb180-1"><a href="analysing-data.html#cb180-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(cat.flowers, <span class="at">pch =</span> <span class="dv">21</span>, <span class="at">bg =</span> <span class="fu">spec2rgb</span>(flowers))</span></code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-85"></span>
<img src="colsci-book_files/figure-html/unnamed-chunk-85-1.png" alt="Flowers in the categorical colourspace of Troje (1993)." width="480" />
<p class="caption">
Figure 3.16: Flowers in the categorical colourspace of Troje (1993).
</p>
</div>
</div>
<div id="segment-classification" class="section level5" number="3.1.4.6.6">
<h5><span class="header-section-number">3.1.4.6.6</span> Segment Classification</h5>
<p>The segment classification analysis <span class="citation">(<a href="#ref-Endler1990" role="doc-biblioref">Endler 1990</a>)</span> does not assume any particular visual system, but instead tries to classify colours in a manner that captures common properties of many vertebrate (and some invertebrate) visual systems. In essence, it breaks down the reflectance spectrum region of interest into four equally-spaced regions, measuring the relative signal along those regions. This approximates a tetrachromatic system with ultraviolet, short, medium, and long-wavelength sensitive photoreceptors.</p>
<p>Though somewhat simplistic, this model captures many of the properties of other, more complex visual models, but without many of the additional assumptions these make. It also provides results in a fairly intuitive colour space, in which the angle corresponds to hue and the distance from the centre corresponds to chroma (Figure below; in fact, variables <code>S5</code> and <code>H4</code> from <code>summary.rspec()</code> are calculated from these relative segments).
Note that, while a segment analysis ranging from 300 or 400 nm to 700 nm corresponds quite closely to the human visual system colour wheel, any wavelength range can be analysed in this way, returning a 360° hue space delimited by the range used.</p>
<p>The segment differences or “opponents” are calculated as:</p>
<p><span class="math display">\[
\begin{split}
LM &amp;= \frac{ R_\lambda \sum_{\lambda={Q4}} R_\lambda - \sum_{\lambda={Q2}} R_\lambda }{\sum_{\lambda={min}}^{max}R_\lambda}\\
MS &amp;= \frac{ R_\lambda \sum_{\lambda={Q3}} R_\lambda - \sum_{\lambda={Q1}} R_\lambda }{\sum_{\lambda={min}}^{max}R_\lambda}
\end{split}
\]</span></p>
<p>Where <span class="math inline">\(Qi\)</span> represent the interquantile distances (e.g. <span class="math inline">\(Q1=ultraviolet\)</span>, <span class="math inline">\(Q2=blue\)</span>, <span class="math inline">\(Q3=green\)</span> and <span class="math inline">\(Q4=red\)</span>)</p>
<p>The segment classification model is obtained through the <code>colspace()</code> argument <code>space = 'segment'</code>, following initial modelling with <code>vismodel()</code>. The example below uses idealized reflectance spectra to illustrate how the avian colour space defined from the segment classification maps to the human colour wheel:</p>
<div class="sourceCode" id="cb181"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb181-1"><a href="analysing-data.html#cb181-1" aria-hidden="true" tabindex="-1"></a>fakedata1 <span class="ot">&lt;-</span> <span class="fu">vapply</span>(</span>
<span id="cb181-2"><a href="analysing-data.html#cb181-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">seq</span>(<span class="dv">100</span>, <span class="dv">500</span>, <span class="at">by =</span> <span class="dv">20</span>),</span>
<span id="cb181-3"><a href="analysing-data.html#cb181-3" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(x) <span class="fu">rowSums</span>(<span class="fu">cbind</span>(</span>
<span id="cb181-4"><a href="analysing-data.html#cb181-4" aria-hidden="true" tabindex="-1"></a>      <span class="fu">dnorm</span>(<span class="dv">300</span><span class="sc">:</span><span class="dv">700</span>, x, <span class="dv">30</span>),</span>
<span id="cb181-5"><a href="analysing-data.html#cb181-5" aria-hidden="true" tabindex="-1"></a>      <span class="fu">dnorm</span>(<span class="dv">300</span><span class="sc">:</span><span class="dv">700</span>, x <span class="sc">+</span> <span class="dv">400</span>, <span class="dv">30</span>)</span>
<span id="cb181-6"><a href="analysing-data.html#cb181-6" aria-hidden="true" tabindex="-1"></a>    )), <span class="fu">numeric</span>(<span class="dv">401</span>)</span>
<span id="cb181-7"><a href="analysing-data.html#cb181-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb181-8"><a href="analysing-data.html#cb181-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb181-9"><a href="analysing-data.html#cb181-9" aria-hidden="true" tabindex="-1"></a><span class="co"># creating idealized specs with varying saturation</span></span>
<span id="cb181-10"><a href="analysing-data.html#cb181-10" aria-hidden="true" tabindex="-1"></a>fakedata2 <span class="ot">&lt;-</span> <span class="fu">vapply</span>(</span>
<span id="cb181-11"><a href="analysing-data.html#cb181-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="dv">500</span>, <span class="dv">300</span>, <span class="dv">150</span>, <span class="dv">105</span>, <span class="dv">75</span>, <span class="dv">55</span>, <span class="dv">40</span>, <span class="dv">30</span>),</span>
<span id="cb181-12"><a href="analysing-data.html#cb181-12" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(x) <span class="fu">dnorm</span>(<span class="dv">300</span><span class="sc">:</span><span class="dv">700</span>, <span class="dv">550</span>, x),</span>
<span id="cb181-13"><a href="analysing-data.html#cb181-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">numeric</span>(<span class="dv">401</span>)</span>
<span id="cb181-14"><a href="analysing-data.html#cb181-14" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb181-15"><a href="analysing-data.html#cb181-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb181-16"><a href="analysing-data.html#cb181-16" aria-hidden="true" tabindex="-1"></a>fakedata1 <span class="ot">&lt;-</span> <span class="fu">as.rspec</span>(<span class="fu">data.frame</span>(<span class="at">wl =</span> <span class="dv">300</span><span class="sc">:</span><span class="dv">700</span>, fakedata1))</span></code></pre></div>
<pre><code>## wavelengths found in column 1</code></pre>
<div class="sourceCode" id="cb183"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb183-1"><a href="analysing-data.html#cb183-1" aria-hidden="true" tabindex="-1"></a>fakedata1 <span class="ot">&lt;-</span> <span class="fu">procspec</span>(fakedata1, <span class="st">&quot;max&quot;</span>)</span></code></pre></div>
<pre><code>## processing options applied:
## Scaling spectra to a maximum value of 1</code></pre>
<div class="sourceCode" id="cb185"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb185-1"><a href="analysing-data.html#cb185-1" aria-hidden="true" tabindex="-1"></a>fakedata2 <span class="ot">&lt;-</span> <span class="fu">as.rspec</span>(<span class="fu">data.frame</span>(<span class="at">wl =</span> <span class="dv">300</span><span class="sc">:</span><span class="dv">700</span>, fakedata2))</span></code></pre></div>
<pre><code>## wavelengths found in column 1</code></pre>
<div class="sourceCode" id="cb187"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb187-1"><a href="analysing-data.html#cb187-1" aria-hidden="true" tabindex="-1"></a>fakedata2 <span class="ot">&lt;-</span> <span class="fu">procspec</span>(fakedata2, <span class="st">&quot;sum&quot;</span>)</span></code></pre></div>
<pre><code>## processing options applied:
## Scaling spectra to a total area of 1</code></pre>
<div class="sourceCode" id="cb189"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb189-1"><a href="analysing-data.html#cb189-1" aria-hidden="true" tabindex="-1"></a>fakedata2 <span class="ot">&lt;-</span> <span class="fu">procspec</span>(fakedata2, <span class="st">&quot;min&quot;</span>)</span></code></pre></div>
<pre><code>## processing options applied:
## Scaling spectra to a minimum value of zero</code></pre>
<div class="sourceCode" id="cb191"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb191-1"><a href="analysing-data.html#cb191-1" aria-hidden="true" tabindex="-1"></a><span class="co"># converting reflectance to percentage</span></span>
<span id="cb191-2"><a href="analysing-data.html#cb191-2" aria-hidden="true" tabindex="-1"></a>fakedata1[, <span class="sc">-</span><span class="dv">1</span>] <span class="ot">&lt;-</span> fakedata1[, <span class="sc">-</span><span class="dv">1</span>] <span class="sc">*</span> <span class="dv">100</span></span>
<span id="cb191-3"><a href="analysing-data.html#cb191-3" aria-hidden="true" tabindex="-1"></a>fakedata2[, <span class="sc">-</span><span class="dv">1</span>] <span class="ot">&lt;-</span> fakedata2[, <span class="sc">-</span><span class="dv">1</span>] <span class="sc">/</span> <span class="fu">max</span>(fakedata2[, <span class="sc">-</span><span class="dv">1</span>]) <span class="sc">*</span> <span class="dv">100</span></span>
<span id="cb191-4"><a href="analysing-data.html#cb191-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb191-5"><a href="analysing-data.html#cb191-5" aria-hidden="true" tabindex="-1"></a><span class="co"># combining and converting to rspec</span></span>
<span id="cb191-6"><a href="analysing-data.html#cb191-6" aria-hidden="true" tabindex="-1"></a>fakedata.c <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">wl =</span> <span class="dv">300</span><span class="sc">:</span><span class="dv">700</span>, fakedata1[, <span class="sc">-</span><span class="dv">1</span>], fakedata2[, <span class="sc">-</span><span class="dv">1</span>])</span>
<span id="cb191-7"><a href="analysing-data.html#cb191-7" aria-hidden="true" tabindex="-1"></a>fakedata.c <span class="ot">&lt;-</span> <span class="fu">as.rspec</span>(fakedata.c)</span></code></pre></div>
<pre><code>## wavelengths found in column 1</code></pre>
<div class="sourceCode" id="cb193"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb193-1"><a href="analysing-data.html#cb193-1" aria-hidden="true" tabindex="-1"></a><span class="co"># segment classification analysis</span></span>
<span id="cb193-2"><a href="analysing-data.html#cb193-2" aria-hidden="true" tabindex="-1"></a>seg.vis <span class="ot">&lt;-</span> <span class="fu">vismodel</span>(fakedata.c, <span class="at">visual =</span> <span class="st">&quot;segment&quot;</span>, <span class="at">achromatic =</span> <span class="st">&quot;all&quot;</span>)</span>
<span id="cb193-3"><a href="analysing-data.html#cb193-3" aria-hidden="true" tabindex="-1"></a>seg.fdc <span class="ot">&lt;-</span> <span class="fu">colspace</span>(seg.vis, <span class="at">space =</span> <span class="st">&quot;segment&quot;</span>)</span>
<span id="cb193-4"><a href="analysing-data.html#cb193-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb193-5"><a href="analysing-data.html#cb193-5" aria-hidden="true" tabindex="-1"></a><span class="co"># plot results</span></span>
<span id="cb193-6"><a href="analysing-data.html#cb193-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(seg.fdc, <span class="at">col =</span> <span class="fu">spec2rgb</span>(fakedata.c))</span></code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-86"></span>
<img src="colsci-book_files/figure-html/unnamed-chunk-86-1.png" alt="Idealized reflectance spectra and their projection on the axes of segment classification" width="384" />
<p class="caption">
Figure 3.17: Idealized reflectance spectra and their projection on the axes of segment classification
</p>
</div>
</div>
<div id="colourspace-distances-with-coldist" class="section level5" number="3.1.4.6.7">
<h5><span class="header-section-number">3.1.4.6.7</span> Colourspace Distances with <code>coldist</code></h5>
<p>Under the colour space framework, colour distances can be calculated simply as Euclidean distances of the relative cone stimulation data, either log-transformed or not, depending on how it was defined. However, these distances cannot be interpreted in terms of JNDs, since no receptor noise is incorporated in the model. Euclidean distances can be computed in <code>R</code> using the <code>coldist()</code> function on the <code>colspace()</code> output:</p>
<div class="sourceCode" id="cb194"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb194-1"><a href="analysing-data.html#cb194-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">coldist</span>(tetra.flowers))</span></code></pre></div>
<pre><code>## Quantum catch are relative, distances may not be meaningfulFALSE</code></pre>
<pre><code>## Calculating unweighted Euclidean distances</code></pre>
<pre><code>##                  patch1              patch2         dS dL
## 1 Goodenia_heterophylla Goodenia_geniculata 0.06695922 NA
## 2 Goodenia_heterophylla   Goodenia_gracilis 0.20467411 NA
## 3 Goodenia_heterophylla    Xyris_operculata 0.05407934 NA
## 4 Goodenia_heterophylla       Eucalyptus_sp 0.01901724 NA
## 5 Goodenia_heterophylla  Faradaya_splendida 0.05027645 NA
## 6 Goodenia_heterophylla  Gaultheria_hispida 0.03897671 NA</code></pre>
<p>Specialised colourspace models such as the colour-hexagon, coc space, CIELab and CIELCh models have their own distance measures, which are returned be default when run through <code>coldist()</code>. Be sure to read the <code>?coldist</code> documentation, and the original publications, to understand what is being returned. As an example, if we run the results of colour-hexagon modelling through <code>coldist()</code>:</p>
<div class="sourceCode" id="cb198"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb198-1"><a href="analysing-data.html#cb198-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Model flower colours according to a honeybee</span></span>
<span id="cb198-2"><a href="analysing-data.html#cb198-2" aria-hidden="true" tabindex="-1"></a>vis.flowers <span class="ot">&lt;-</span> <span class="fu">vismodel</span>(flowers, <span class="at">visual =</span> <span class="st">&quot;apis&quot;</span>, <span class="at">qcatch =</span> <span class="st">&quot;Ei&quot;</span>, <span class="at">relative =</span> <span class="cn">FALSE</span>, <span class="at">vonkries =</span> <span class="cn">TRUE</span>, <span class="at">achromatic =</span> <span class="st">&quot;l&quot;</span>, <span class="at">bkg =</span> <span class="st">&quot;green&quot;</span>)</span>
<span id="cb198-3"><a href="analysing-data.html#cb198-3" aria-hidden="true" tabindex="-1"></a>hex.flowers <span class="ot">&lt;-</span> <span class="fu">colspace</span>(vis.flowers, <span class="at">space =</span> <span class="st">&quot;hexagon&quot;</span>)</span>
<span id="cb198-4"><a href="analysing-data.html#cb198-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb198-5"><a href="analysing-data.html#cb198-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Estimate colour distances. No need to specify relative receptor densities, noise etc.,</span></span>
<span id="cb198-6"><a href="analysing-data.html#cb198-6" aria-hidden="true" tabindex="-1"></a><span class="co"># which only apply in the case of receptor-noise modelling</span></span>
<span id="cb198-7"><a href="analysing-data.html#cb198-7" aria-hidden="true" tabindex="-1"></a>dist.flowers <span class="ot">&lt;-</span> <span class="fu">coldist</span>(hex.flowers)</span></code></pre></div>
<pre><code>## Calculating unweighted Euclidean distances</code></pre>
<div class="sourceCode" id="cb200"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb200-1"><a href="analysing-data.html#cb200-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(dist.flowers)</span></code></pre></div>
<pre><code>##                  patch1              patch2        dS dL
## 1 Goodenia_heterophylla Goodenia_geniculata 0.3207265 NA
## 2 Goodenia_heterophylla   Goodenia_gracilis 0.6870945 NA
## 3 Goodenia_heterophylla    Xyris_operculata 0.3025298 NA
## 4 Goodenia_heterophylla       Eucalyptus_sp 0.1111375 NA
## 5 Goodenia_heterophylla  Faradaya_splendida 0.2324927 NA
## 6 Goodenia_heterophylla  Gaultheria_hispida 0.2280836 NA</code></pre>
<p>the chromatic contrasts <code>dS</code> and achromatic contrasts <code>dL</code> are expressed as Euclidean distances in the hexagon, known as ‘hexagon units’ <span class="citation">(<a href="#ref-Chittka1992" role="doc-biblioref">Chittka 1992</a>)</span>. If we had instead used the colour-opponent coding space the units would have been city-bloc distances, while in the CIELab model the distances would be derived from the CIE colour-distance formula (2000).</p>
</div>
<div id="distances-in-n-dimensions" class="section level5" number="3.1.4.6.8">
<h5><span class="header-section-number">3.1.4.6.8</span> Distances in N-Dimensions</h5>
<p>The <code>coldist()</code> function is no longer limited to di-, tri- or tetrachromatic visual systems. The function has been generalised, and can now calculate colour distances for n-dimensional visual phenotypes. This means there is no limit on the dimensionality of visual systems that may be input, which may prove useful for modelling nature’s extremes (e.g. butterflies, mantis shrimp) or for simulation-based work. Naturally, since these calculations aren’t largely implemented elsewhere, we recommend caution and validation of results prior to publication.</p>
<div class="sourceCode" id="cb202"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb202-1"><a href="analysing-data.html#cb202-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create an arbitrary visual phenotype with 10 photoreceptors</span></span>
<span id="cb202-2"><a href="analysing-data.html#cb202-2" aria-hidden="true" tabindex="-1"></a>fakemantisshrimp <span class="ot">&lt;-</span> <span class="fu">sensmodel</span>(<span class="fu">c</span>(<span class="dv">325</span>, <span class="dv">350</span>, <span class="dv">400</span>, <span class="dv">425</span>, <span class="dv">450</span>, <span class="dv">500</span>, <span class="dv">550</span>, <span class="dv">600</span>, <span class="dv">650</span>, <span class="dv">700</span>), <span class="at">beta =</span> <span class="cn">FALSE</span>, <span class="at">integrate =</span> <span class="cn">FALSE</span>)</span>
<span id="cb202-3"><a href="analysing-data.html#cb202-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb202-4"><a href="analysing-data.html#cb202-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to percentages, just to colour the plot</span></span>
<span id="cb202-5"><a href="analysing-data.html#cb202-5" aria-hidden="true" tabindex="-1"></a>fakemantisshrimp.colours <span class="ot">&lt;-</span> fakemantisshrimp <span class="sc">*</span> <span class="dv">100</span></span>
<span id="cb202-6"><a href="analysing-data.html#cb202-6" aria-hidden="true" tabindex="-1"></a>fakemantisshrimp.colours[, <span class="st">&quot;wl&quot;</span>] <span class="ot">&lt;-</span> fakemantisshrimp[, <span class="st">&quot;wl&quot;</span>]</span>
<span id="cb202-7"><a href="analysing-data.html#cb202-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb202-8"><a href="analysing-data.html#cb202-8" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(fakemantisshrimp, <span class="at">col =</span> <span class="fu">spec2rgb</span>(fakemantisshrimp.colours), <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">ylab =</span> <span class="st">&quot;Absorbance&quot;</span>)</span></code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-89"></span>
<img src="colsci-book_files/figure-html/unnamed-chunk-89-1.png" alt="Visual system of a pretend mantis shrimp with 10 cones" width="384" />
<p class="caption">
Figure 3.18: Visual system of a pretend mantis shrimp with 10 cones
</p>
</div>
<div class="sourceCode" id="cb203"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb203-1"><a href="analysing-data.html#cb203-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Run visual model and calculate colour distances</span></span>
<span id="cb203-2"><a href="analysing-data.html#cb203-2" aria-hidden="true" tabindex="-1"></a>vm.fms <span class="ot">&lt;-</span> <span class="fu">vismodel</span>(flowers, <span class="at">visual =</span> fakemantisshrimp, <span class="at">relative =</span> <span class="cn">FALSE</span>, <span class="at">achromatic =</span> <span class="cn">FALSE</span>)</span>
<span id="cb203-3"><a href="analysing-data.html#cb203-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb203-4"><a href="analysing-data.html#cb203-4" aria-hidden="true" tabindex="-1"></a>JND.fms <span class="ot">&lt;-</span> <span class="fu">coldist</span>(vm.fms, <span class="at">n =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">4</span>, <span class="dv">5</span>, <span class="dv">5</span>))</span></code></pre></div>
<pre><code>## Calculating noise-weighted Euclidean distances</code></pre>
<div class="sourceCode" id="cb205"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb205-1"><a href="analysing-data.html#cb205-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(JND.fms)</span></code></pre></div>
<pre><code>##                  patch1              patch2        dS dL
## 1 Goodenia_heterophylla Goodenia_geniculata 16.297557 NA
## 2 Goodenia_heterophylla   Goodenia_gracilis 48.390309 NA
## 3 Goodenia_heterophylla    Xyris_operculata 30.760482 NA
## 4 Goodenia_heterophylla       Eucalyptus_sp  6.044184 NA
## 5 Goodenia_heterophylla  Faradaya_splendida 15.901435 NA
## 6 Goodenia_heterophylla  Gaultheria_hispida 11.188246 NA</code></pre>
</div>
</div>
</div>
</div>
<div id="analysing-spatial-data" class="section level2" number="3.2">
<h2><span class="header-section-number">3.2</span> Analysing Spatial Data</h2>
<p><code>pavo</code> now allows for the combined analysis of spectral data and spatial data. Note that ‘spatial’ data of course include images, but may also draw on spatially-sampled <em>spectra</em>, as explored below. Thus while much of the below deals with image-based processing and analysis, images are not necessary for many spatial analyses and indeed may perform worse than spectrally-based analyses in some cases. Further, <strong>image data should seldom be used without input from spectral measurements</strong>. Even in the simple butterfly example below, users should first validate the presence and number of discrete colours in a patch via spectral measurement and visual modelling (i.e. the ‘spectral’ information), which is then augmented by image-based clustering of these <em>a priori</em> identified patches to extract the ‘spatial’ information. Using uncalibrated images alone for inferences about colour pattern perception may well yield unreliable results, given the radical differences between human (camera) and non-human animal perception.</p>
<div id="image-based-colour-classification" class="section level3" number="3.2.1">
<h3><span class="header-section-number">3.2.1</span> Image-based colour classification</h3>
<p>We first import our images with <code>getimg()</code>, which functions in a similar manner to <code>getspec()</code>. Since we are importing more than one image, we can simply point the function to the folder containing the images, and it will use parallel processing (where possible) to import any jpg, bmp, or png images in the folder. As with the raw spectra, these images are also available at the package repository [here][data-location], but are included in the package installation for convenience.</p>
<div class="sourceCode" id="cb207"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb207-1"><a href="analysing-data.html#cb207-1" aria-hidden="true" tabindex="-1"></a>butterflies <span class="ot">&lt;-</span> <span class="fu">getimg</span>(<span class="fu">system.file</span>(<span class="st">&quot;testdata/images/&quot;</span>, <span class="at">package =</span> <span class="st">&#39;pavo&#39;</span>))</span></code></pre></div>
<pre><code>## 2 files found; importing images.</code></pre>
<p>The segmentation of RGB images into discrete colour-classes is an useful step on the road to several analyses, and this is carried out in <code>pavo</code> via <code>classify()</code>. The function currently implements k-means clustering which, and offers some flexibility in the way in which it is carried out. In the simplest case, we can specify an image or image(s) as well as the number of discrete colours in each, and allow the k-means clustering algorithm to do its work. <strong>Note that the specification of <em>k</em>, the number of discrete colours in an image, is a crucial step to be estimated <em>a priori</em> beforehand</strong>, typically through the use of visual modelling to estimate the number of discriminable colours in an ‘objective’ manner. There are a suite of possible ways in which this may be achieved, though, with one detailed below as part of example 2.</p>
<p>Since we have two images, with (arguably) four and three colours each, including the white background, we can specify this information using <code>kcols</code>, before visualising the results with <code>summary</code>. The <code>summary</code> function, with <code>plot = TRUE</code>, is also particularly useful for understanding the colour-class labels, or ID’s, assigned to each colour, as well as visualising the accuracy of the classification process, which may be enhanced by specifying strongly-contrasting ‘false’ colours for plotting through the <code>col</code> argument.</p>
<div class="sourceCode" id="cb209"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb209-1"><a href="analysing-data.html#cb209-1" aria-hidden="true" tabindex="-1"></a>butterflies_class <span class="ot">&lt;-</span> <span class="fu">classify</span>(butterflies, <span class="at">kcols =</span> <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">3</span>))</span></code></pre></div>
<pre><code>## Image classification in progress...</code></pre>
<div class="sourceCode" id="cb211"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb211-1"><a href="analysing-data.html#cb211-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Note that we could simply feed the list of images to summary, rather than</span></span>
<span id="cb211-2"><a href="analysing-data.html#cb211-2" aria-hidden="true" tabindex="-1"></a><span class="co"># specifying individual images, and they would progress automatically</span></span>
<span id="cb211-3"><a href="analysing-data.html#cb211-3" aria-hidden="true" tabindex="-1"></a><span class="co"># with user input.</span></span>
<span id="cb211-4"><a href="analysing-data.html#cb211-4" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(butterflies_class[[<span class="dv">2</span>]], <span class="at">plot =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-92"></span>
<img src="colsci-book_files/figure-html/unnamed-chunk-92-1.png" alt="The k-means classified images of our butterflies, along with their identified colour palettes" width="480" />
<p class="caption">
Figure 3.19: The k-means classified images of our butterflies, along with their identified colour palettes
</p>
</div>
<p>There are other options available, which may be of particular use when the classification process is not producing the results that might be expected. If analysing similar images with homologous colour patterns (e.g. individuals within a species), we could specify a single ‘reference’ image in a list to be initially classified, using <code>refID</code>, with the RGB centres identified in this reference image used as pre-specified cluster centres in subsequent images. This helps to ensure that homologous pattern elements are reliably identified and classified between images. The <code>classify()</code> function also offers an interactive option, via <code>interactive = TRUE</code>, wherein we can interactively click on representative colours within an image (or images individually), with the RGB values then used as cluster centres. This is particularly useful when images are being erroneously classified, since the standard k-means process randomly selects initial centres and so may (by chance) miss smaller patches, or confuse similar colours. By specifying the initial RGB centres interactively, it can be easier for the process to distinguish or ‘pull apart’ colours in such circumstances. Ultimately, there are several combinations of options available during classification, which are also documented in the help file. For our butterflies, for example, we could use any of the following (though some would be very ill-advised):</p>
<div class="sourceCode" id="cb212"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb212-1"><a href="analysing-data.html#cb212-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Automatic classification.</span></span>
<span id="cb212-2"><a href="analysing-data.html#cb212-2" aria-hidden="true" tabindex="-1"></a>butterflies_class <span class="ot">&lt;-</span> <span class="fu">classify</span>(butterflies, <span class="at">kcols =</span> <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">3</span>))</span></code></pre></div>
<pre><code>## Image classification in progress...</code></pre>
<div class="sourceCode" id="cb214"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb214-1"><a href="analysing-data.html#cb214-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Automatic classification using a reference image, identified by name.</span></span>
<span id="cb214-2"><a href="analysing-data.html#cb214-2" aria-hidden="true" tabindex="-1"></a>butterflies_class <span class="ot">&lt;-</span> <span class="fu">classify</span>(butterflies, <span class="at">refID =</span> <span class="st">&#39;h_melpomene&#39;</span>, <span class="at">kcols =</span> <span class="dv">3</span>)</span></code></pre></div>
<pre><code>## Image classification in progress...</code></pre>
<div class="sourceCode" id="cb216"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb216-1"><a href="analysing-data.html#cb216-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Classification using interactively-specified centres for each image, with no</span></span>
<span id="cb216-2"><a href="analysing-data.html#cb216-2" aria-hidden="true" tabindex="-1"></a><span class="co"># need to specify kcols (since it will be inferred from the numbers of colours selected)</span></span>
<span id="cb216-3"><a href="analysing-data.html#cb216-3" aria-hidden="true" tabindex="-1"></a>butterflies_class <span class="ot">&lt;-</span> <span class="fu">classify</span>(butterflies, <span class="at">interactive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb216-4"><a href="analysing-data.html#cb216-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb216-5"><a href="analysing-data.html#cb216-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Classification using interactively-specified centres for the single reference image</span></span>
<span id="cb216-6"><a href="analysing-data.html#cb216-6" aria-hidden="true" tabindex="-1"></a><span class="co"># (here, the first in the list). We could also specify reference image using it&#39;s name, </span></span>
<span id="cb216-7"><a href="analysing-data.html#cb216-7" aria-hidden="true" tabindex="-1"></a><span class="co"># as above.</span></span>
<span id="cb216-8"><a href="analysing-data.html#cb216-8" aria-hidden="true" tabindex="-1"></a>butterflies_class <span class="ot">&lt;-</span> <span class="fu">classify</span>(butterflies, <span class="at">refID =</span> <span class="dv">1</span>, <span class="at">interactive =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
</div>
<div id="adjacency-and-boundary-strength-analyses-and-overall-pattern-contrasts" class="section level3" number="3.2.2">
<h3><span class="header-section-number">3.2.2</span> Adjacency and Boundary Strength Analyses, and Overall Pattern Contrasts</h3>
<p>The adjacency <span class="citation">(<a href="#ref-Endler2012" role="doc-biblioref">Endler 2012</a>)</span> and boundary strength <span class="citation">(<a href="#ref-Endler2018" role="doc-biblioref">Endler, Cole, and Kranz 2018</a>)</span> analyses offer a powerful method for quantifying various features of colour pattern geometry. Briefly, the process entails classifying grid-sampled locations within a visual scene into a number of discrete colour classes. This may be achieved using <code>classify()</code> as described above before sampling the now-classified image, and/or by taking grid-sampled reflectance spectra across a visual scene, visually modelling and clustering the results <span class="citation">(as per <span>‘method 1’</span> in <a href="#ref-Endler2012" role="doc-biblioref">Endler 2012</a>, example below)</span>. The column-wise and row-wise colour-class transitions between adjacent points within this grid are then tallied, and from this a suite of summary statistics on pattern structure — from simple colour proportions, through to colour diversity and pattern complexity — are estimated. If the colour ‘distance’ between adjacent colour classes is known, such as might be estimated using spectral data for receptor-noise modelling via <code>vismodel()</code> and <code>coldist()</code>, then this can also be incorporated to derive several measures of the salience of patch boundaries <span class="citation">(<a href="#ref-Endler2018" role="doc-biblioref">Endler, Cole, and Kranz 2018</a>)</span>. Finally, if estimates of hue, saturation, and/or brightness are known for each discrete colour in a pattern, then these can be included through the <code>hsl</code> argument for general, area-weighted measures of overall pattern contrast <span class="citation">(<a href="#ref-Endler2005" role="doc-biblioref">Endler and Mielke 2005</a>)</span>. In <code>pavo</code>, this is all carried out through the <code>adjacent()</code> function, which takes several arguments that should be well understood.</p>
<div id="example-1-simple-mounted-butterflies" class="section level4" number="3.2.2.1">
<h4><span class="header-section-number">3.2.2.1</span> Example 1 (simple): mounted butterflies</h4>
<p>We’ll begin with a simple example using mounted butterflies, by first loading and classifying our images in the manner outlined above. For the adjacency analysis proper, we’ll specify a full x-dimension scale of 200 mm using <code>xscale</code> (which could be ignored if we specify scales using <code>procimg()</code>), and a sampling density of 200, for a 200 x 200 grid. Since the white background of our images is fake and of no interest, we will also exclude it by specifying the colour-class ID that corresponds to it (for more complex backgrounds, we could instead use <code>procimg()</code>, as above, and ignore this argument).</p>
<div class="sourceCode" id="cb217"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb217-1"><a href="analysing-data.html#cb217-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load up our images</span></span>
<span id="cb217-2"><a href="analysing-data.html#cb217-2" aria-hidden="true" tabindex="-1"></a>butterflies <span class="ot">&lt;-</span> <span class="fu">getimg</span>(<span class="fu">system.file</span>(<span class="st">&quot;testdata/images/&quot;</span>, <span class="at">package =</span> <span class="st">&#39;pavo&#39;</span>))</span></code></pre></div>
<pre><code>## 2 files found; importing images.</code></pre>
<div class="sourceCode" id="cb219"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb219-1"><a href="analysing-data.html#cb219-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Automatically classify discrete colour patches (as confirmed by spectral modelling)</span></span>
<span id="cb219-2"><a href="analysing-data.html#cb219-2" aria-hidden="true" tabindex="-1"></a><span class="co"># using k-means clustering, with values of &#39;k&#39; that have been validated a priori.</span></span>
<span id="cb219-3"><a href="analysing-data.html#cb219-3" aria-hidden="true" tabindex="-1"></a>butterflies_class <span class="ot">&lt;-</span> <span class="fu">classify</span>(butterflies, <span class="at">kcols =</span> <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">3</span>))</span></code></pre></div>
<pre><code>## Image classification in progress...</code></pre>
<div class="sourceCode" id="cb221"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb221-1"><a href="analysing-data.html#cb221-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Run the adjacency analysis, subsampling the image in a 200x200 grid, and excluding</span></span>
<span id="cb221-2"><a href="analysing-data.html#cb221-2" aria-hidden="true" tabindex="-1"></a><span class="co"># the white background.</span></span>
<span id="cb221-3"><a href="analysing-data.html#cb221-3" aria-hidden="true" tabindex="-1"></a>butterflies_adj <span class="ot">&lt;-</span> <span class="fu">adjacent</span>(butterflies_class, <span class="at">xscale =</span> <span class="dv">200</span>, <span class="at">xpts =</span> <span class="dv">200</span>, <span class="at">bkgID =</span> <span class="dv">1</span>)</span>
<span id="cb221-4"><a href="analysing-data.html#cb221-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb221-5"><a href="analysing-data.html#cb221-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a look</span></span>
<span id="cb221-6"><a href="analysing-data.html#cb221-6" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(butterflies_adj)</span></code></pre></div>
<pre><code>##             k     N n_off       p_1       p_2       p_3        p_4     q_1_1
## h_melpomene 4 54064  5111 0.5833824 0.3408824 0.0406250 0.03511029 0.5649970
## papilio     3 63241  4371 0.1995597 0.5609434 0.2394969         NA 0.1726728
##                   q_1_2     q_2_2      q_2_3      q_3_3      q_1_4      q_2_4
## h_melpomene 0.018644569 0.3071175 0.01825614 0.02683856 0.01294762 0.03486608
## papilio     0.006941699 0.5485682 0.01315602 0.20964248         NA         NA
##                   q_3_4       q_4_4     t_1_2     t_2_3     t_1_4     t_2_4
## h_melpomene 0.009821693 0.006510802 0.1972217 0.1931129 0.1369595 0.3688124
## papilio              NA          NA 0.1004347 0.1903455        NA        NA
##                 t_3_4      q_1_3     t_1_3          m        m_r        m_c
## h_melpomene 0.1038936         NA        NA 0.09395221 0.08779412 0.10011029
## papilio            NA 0.04901883 0.7092199 0.06868910 0.07805031 0.05932789
##                     A       Sc       St        Jc        Jt         B        Rt
## h_melpomene 0.8769739 2.176663 4.136277 0.5441657 0.6893795 0.8839757 0.5665327
## papilio     1.3155754 2.428125 1.820461 0.8093752 0.6068204 0.2150538 0.5493113
##             Rab m_dS s_dS cv_dS m_dL s_dL cv_dL m_hue s_hue var_hue m_sat s_sat
## h_melpomene   1   NA   NA    NA   NA   NA    NA    NA    NA      NA    NA    NA
## papilio       1   NA   NA    NA   NA   NA    NA    NA    NA      NA    NA    NA
##             cv_sat m_lum s_lum cv_lum
## h_melpomene     NA    NA    NA     NA
## papilio         NA    NA    NA     NA</code></pre>
<p>The output is a suite of statistics describing colour pattern structure and geometry in a tidy data frame that, since we fed it a list of images, contains one row for each of our samples. As always, the function’s help file (<code>?adjacent</code>) summarises the meaning of each output variable, and the original references listed therein contain the full discussion. Note that some of the output is <code>NA</code>, which might occur when the focal statistic simply makes no sense for the image in question (such as the transition frequency between colour categories 3 and 4, for an image that contains only 3 colour categories). Others may result when we do not give <code>adjacent()</code> enough information, such as the colour-distances required to calculate the boundary-strength statistics, or the colourmetrics required for overall pattern contrast. If we do give the function this information, we can see that those final variables now contain sensible output.</p>
<div class="sourceCode" id="cb223"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb223-1"><a href="analysing-data.html#cb223-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a fake matrix of pairwise colour- and luminance distances between all</span></span>
<span id="cb223-2"><a href="analysing-data.html#cb223-2" aria-hidden="true" tabindex="-1"></a><span class="co"># colour patten elements, as might be attained through visual modelling of spectral data.</span></span>
<span id="cb223-3"><a href="analysing-data.html#cb223-3" aria-hidden="true" tabindex="-1"></a>distances <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">c1 =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">2</span>),</span>
<span id="cb223-4"><a href="analysing-data.html#cb223-4" aria-hidden="true" tabindex="-1"></a>                        <span class="at">c2 =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">3</span>),</span>
<span id="cb223-5"><a href="analysing-data.html#cb223-5" aria-hidden="true" tabindex="-1"></a>                        <span class="at">dS =</span> <span class="fu">c</span>(<span class="fl">10.6</span>, <span class="fl">5.1</span>, <span class="fl">4.4</span>),</span>
<span id="cb223-6"><a href="analysing-data.html#cb223-6" aria-hidden="true" tabindex="-1"></a>                        <span class="at">dL =</span> <span class="fu">c</span>(<span class="fl">1.1</span>, <span class="fl">2.5</span>, <span class="fl">3.2</span>))</span>
<span id="cb223-7"><a href="analysing-data.html#cb223-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb223-8"><a href="analysing-data.html#cb223-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a look</span></span>
<span id="cb223-9"><a href="analysing-data.html#cb223-9" aria-hidden="true" tabindex="-1"></a>distances</span></code></pre></div>
<pre><code>##   c1 c2   dS  dL
## 1  1  2 10.6 1.1
## 2  1  3  5.1 2.5
## 3  2  3  4.4 3.2</code></pre>
<div class="sourceCode" id="cb225"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb225-1"><a href="analysing-data.html#cb225-1" aria-hidden="true" tabindex="-1"></a><span class="co"># And our fake hue angles (in radians), saturation, and luminance values, for each</span></span>
<span id="cb225-2"><a href="analysing-data.html#cb225-2" aria-hidden="true" tabindex="-1"></a><span class="co"># colour pattern element</span></span>
<span id="cb225-3"><a href="analysing-data.html#cb225-3" aria-hidden="true" tabindex="-1"></a>hsl_vals <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">patch =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>,</span>
<span id="cb225-4"><a href="analysing-data.html#cb225-4" aria-hidden="true" tabindex="-1"></a>                       <span class="at">hue =</span> <span class="fu">c</span>(<span class="fl">1.2</span>, <span class="fl">2.2</span>, <span class="fl">1.6</span>),</span>
<span id="cb225-5"><a href="analysing-data.html#cb225-5" aria-hidden="true" tabindex="-1"></a>                       <span class="at">lum =</span> <span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">5</span>, <span class="dv">7</span>),</span>
<span id="cb225-6"><a href="analysing-data.html#cb225-6" aria-hidden="true" tabindex="-1"></a>                       <span class="at">sat =</span> <span class="fu">c</span>(<span class="fl">3.5</span>, <span class="fl">1.1</span>, <span class="fl">6.3</span>))</span>
<span id="cb225-7"><a href="analysing-data.html#cb225-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb225-8"><a href="analysing-data.html#cb225-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a look</span></span>
<span id="cb225-9"><a href="analysing-data.html#cb225-9" aria-hidden="true" tabindex="-1"></a>hsl_vals</span></code></pre></div>
<pre><code>##   patch hue lum sat
## 1     1 1.2  10 3.5
## 2     2 2.2   5 1.1
## 3     3 1.6   7 6.3</code></pre>
<div class="sourceCode" id="cb227"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb227-1"><a href="analysing-data.html#cb227-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Now feed this information into the adjacency analysis using the &#39;less-colourful&#39;</span></span>
<span id="cb227-2"><a href="analysing-data.html#cb227-2" aria-hidden="true" tabindex="-1"></a><span class="co"># of our two images, for convenience (though this could be readily extended to</span></span>
<span id="cb227-3"><a href="analysing-data.html#cb227-3" aria-hidden="true" tabindex="-1"></a><span class="co"># include a list of images along with a list of distances and hsl values)</span></span>
<span id="cb227-4"><a href="analysing-data.html#cb227-4" aria-hidden="true" tabindex="-1"></a><span class="fu">adjacent</span>(butterflies_class[[<span class="dv">2</span>]], <span class="at">xscale =</span> <span class="dv">200</span>, <span class="at">xpts =</span> <span class="dv">200</span>, <span class="at">bkgID =</span> <span class="dv">1</span>, <span class="at">coldists =</span> distances, <span class="at">hsl =</span> hsl_vals)</span></code></pre></div>
<pre><code>##         k     N n_off       p_1       p_2       p_3     q_1_1       q_1_2
## papilio 3 63241  4371 0.1995597 0.5609434 0.2394969 0.1726728 0.006941699
##             q_2_2      q_1_3      q_2_3     q_3_3     t_1_2     t_1_3     t_2_3
## papilio 0.5485682 0.04901883 0.01315602 0.2096425 0.1004347 0.7092199 0.1903455
##                 m        m_r        m_c        A       Sc       St        Jc
## papilio 0.0686891 0.07805031 0.05932789 1.315575 2.428125 1.820461 0.8093752
##                Jt         B        Rt Rab     m_dS    s_dS     cv_dS     m_dL
## papilio 0.6068204 0.2150538 0.5493113   1 5.519149 2.56091 0.4640044 2.492633
##             s_dL     cv_dL    m_hue     s_hue   var_hue    m_sat    s_sat
## papilio 0.802252 0.3218492 1.663715 0.4154308 0.0826731 2.824327 2.812416
##            cv_sat    m_lum    s_lum    cv_lum
## papilio 0.9957825 6.476792 2.530441 0.3906936</code></pre>
</div>
<div id="example-2-complex-whole-visual-environments" class="section level4" number="3.2.2.2">
<h4><span class="header-section-number">3.2.2.2</span> Example 2 (complex): whole visual environments</h4>
<p>Of course colour patterns are not always so simple, and we may want to analyse the spatial geometry of more complex or subtle scenes. In such cases, it may be preferable to largely forego image-based clustering and classification altogether and instead use spectral data directly in an adjacency-style analysis <span class="citation">(as per <span>‘method 1’</span> in <a href="#ref-Endler2012" role="doc-biblioref">Endler 2012</a>)</span>. One such approach would have us take spectral measurements across a patch in an evenly-spaced grid</p>
<div class="sourceCode" id="cb229"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb229-1"><a href="analysing-data.html#cb229-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load up our image of a camouflaged lizard</span></span>
<span id="cb229-2"><a href="analysing-data.html#cb229-2" aria-hidden="true" tabindex="-1"></a>lizard <span class="ot">&lt;-</span> <span class="fu">getimg</span>(<span class="fu">system.file</span>(<span class="st">&quot;testdata/images/vig/&quot;</span>, <span class="at">package =</span> <span class="st">&#39;pavo&#39;</span>))</span></code></pre></div>
<pre><code>## 1 files found; importing images.</code></pre>
<div class="sourceCode" id="cb231"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb231-1"><a href="analysing-data.html#cb231-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a look at it</span></span>
<span id="cb231-2"><a href="analysing-data.html#cb231-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(lizard)</span>
<span id="cb231-3"><a href="analysing-data.html#cb231-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb231-4"><a href="analysing-data.html#cb231-4" aria-hidden="true" tabindex="-1"></a><span class="co"># And overlay a simple &#39;sampling grid&#39; that might guide us as to where we should collect</span></span>
<span id="cb231-5"><a href="analysing-data.html#cb231-5" aria-hidden="true" tabindex="-1"></a><span class="co"># our reflectance measurements. Note that the grid here is likely too coarse to capture the necessary detail,</span></span>
<span id="cb231-6"><a href="analysing-data.html#cb231-6" aria-hidden="true" tabindex="-1"></a><span class="co"># though the sampling density is something to be closely considered on a case-by-case basis</span></span>
<span id="cb231-7"><a href="analysing-data.html#cb231-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(lizard)</span>
<span id="cb231-8"><a href="analysing-data.html#cb231-8" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(<span class="fu">expand.grid</span>(<span class="fu">seq</span>(<span class="dv">0</span>, <span class="fu">dim</span>(lizard)[<span class="dv">1</span>], <span class="dv">25</span>), <span class="fu">seq</span>(<span class="dv">0</span>, <span class="fu">dim</span>(lizard)[<span class="dv">1</span>], <span class="dv">25</span>)), <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">col =</span> <span class="st">&#39;red&#39;</span>)</span></code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-97"></span>
<img src="colsci-book_files/figure-html/unnamed-chunk-97-1.png" alt="A cryptic lizard, along with a possible sampling grid for spectral measurement." width="45%" />
<p class="caption">
Figure 3.20: A cryptic lizard, along with a possible sampling grid for spectral measurement.
</p>
</div>
<p>For the sake of this example we’ll simulate 100 reflectance spectra, as though we had collected them following the above grid. We’ll do this by creating four very similar groups of of 25 spectra, which will also allow us to get a sense of how our clustering methods perform. Of course real spectra will be more variable, but the same principles apply.</p>
<div class="sourceCode" id="cb232"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb232-1"><a href="analysing-data.html#cb232-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">12352</span>)  <span class="co"># For reprodicubility</span></span>
<span id="cb232-2"><a href="analysing-data.html#cb232-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb232-3"><a href="analysing-data.html#cb232-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate some fake spectra that were grid-sampled across our scene</span></span>
<span id="cb232-4"><a href="analysing-data.html#cb232-4" aria-hidden="true" tabindex="-1"></a>fakescene <span class="ot">&lt;-</span> <span class="fu">cbind</span>(</span>
<span id="cb232-5"><a href="analysing-data.html#cb232-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">do.call</span>(cbind, <span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">25</span>, <span class="cf">function</span>(x) <span class="fu">dnorm</span>(<span class="dv">300</span><span class="sc">:</span><span class="dv">700</span>, <span class="fu">runif</span>(<span class="dv">1</span>, <span class="dv">320</span>, <span class="dv">330</span>), <span class="dv">20</span>))),</span>
<span id="cb232-6"><a href="analysing-data.html#cb232-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">do.call</span>(cbind, <span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">25</span>, <span class="cf">function</span>(x) <span class="fu">dnorm</span>(<span class="dv">300</span><span class="sc">:</span><span class="dv">700</span>, <span class="fu">runif</span>(<span class="dv">1</span>, <span class="dv">440</span>, <span class="dv">450</span>), <span class="dv">20</span>))),</span>
<span id="cb232-7"><a href="analysing-data.html#cb232-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">do.call</span>(cbind, <span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">25</span>, <span class="cf">function</span>(x) <span class="fu">dnorm</span>(<span class="dv">300</span><span class="sc">:</span><span class="dv">700</span>, <span class="fu">runif</span>(<span class="dv">1</span>, <span class="dv">550</span>, <span class="dv">560</span>), <span class="dv">20</span>))),</span>
<span id="cb232-8"><a href="analysing-data.html#cb232-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">do.call</span>(cbind, <span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">25</span>, <span class="cf">function</span>(x) <span class="fu">dnorm</span>(<span class="dv">300</span><span class="sc">:</span><span class="dv">700</span>, <span class="fu">runif</span>(<span class="dv">1</span>, <span class="dv">650</span>, <span class="dv">660</span>), <span class="dv">20</span>)))</span>
<span id="cb232-9"><a href="analysing-data.html#cb232-9" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb232-10"><a href="analysing-data.html#cb232-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb232-11"><a href="analysing-data.html#cb232-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert them to rspce objects</span></span>
<span id="cb232-12"><a href="analysing-data.html#cb232-12" aria-hidden="true" tabindex="-1"></a>fakescene <span class="ot">&lt;-</span> <span class="fu">as.rspec</span>(<span class="fu">data.frame</span>(<span class="at">wl =</span> <span class="dv">300</span><span class="sc">:</span><span class="dv">700</span>, fakescene))</span></code></pre></div>
<p>Now instead of calling <code>cluster()</code> on the image and arbitrarily selecting <code>kcols</code> (which would likely be quite inaccurate due to the complexity of the scene), we can visually model the spectra, calculate the colour-distances between them, and cluster the results into a data-image that can be fed directly into <code>adjacent()</code>. Below we’ll use a centroid-based ‘hartigan’ clustering method via the package <code>NbClust</code>, but there are numerous possible approaches that might be used, which should be closely considered with reference to the original methods’ publications.</p>
<div class="sourceCode" id="cb233"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb233-1"><a href="analysing-data.html#cb233-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Visually model our spectra in a tetrahedral model of the blue tit, specifying</span></span>
<span id="cb233-2"><a href="analysing-data.html#cb233-2" aria-hidden="true" tabindex="-1"></a><span class="co"># relative = FALSE so that we can estimate noise-calibrated distances.</span></span>
<span id="cb233-3"><a href="analysing-data.html#cb233-3" aria-hidden="true" tabindex="-1"></a>vis.fakescene <span class="ot">&lt;-</span> <span class="fu">vismodel</span>(fakescene, <span class="at">visual =</span> <span class="st">&#39;bluetit&#39;</span>, <span class="at">relative =</span> <span class="cn">FALSE</span>, <span class="at">scale =</span> <span class="dv">10000</span>)</span>
<span id="cb233-4"><a href="analysing-data.html#cb233-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb233-5"><a href="analysing-data.html#cb233-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate noise-weighted distances, before converting them to xyz coordinates</span></span>
<span id="cb233-6"><a href="analysing-data.html#cb233-6" aria-hidden="true" tabindex="-1"></a><span class="co"># in noise-corrected colourspace.</span></span>
<span id="cb233-7"><a href="analysing-data.html#cb233-7" aria-hidden="true" tabindex="-1"></a>jnd.fakescene <span class="ot">&lt;-</span> <span class="fu">jnd2xyz</span>(<span class="fu">coldist</span>(vis.fakescene))</span>
<span id="cb233-8"><a href="analysing-data.html#cb233-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb233-9"><a href="analysing-data.html#cb233-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a distance matrix, for clustering</span></span>
<span id="cb233-10"><a href="analysing-data.html#cb233-10" aria-hidden="true" tabindex="-1"></a>jnd.mat <span class="ot">&lt;-</span> <span class="fu">dist</span>(jnd.fakescene, <span class="at">method =</span> <span class="st">&#39;euclidean&#39;</span>, <span class="at">diag =</span> <span class="cn">FALSE</span>)</span>
<span id="cb233-11"><a href="analysing-data.html#cb233-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb233-12"><a href="analysing-data.html#cb233-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Zero all distance &lt; 1 (&#39;just noticeable distance&#39;, in this model), which corresponds</span></span>
<span id="cb233-13"><a href="analysing-data.html#cb233-13" aria-hidden="true" tabindex="-1"></a><span class="co"># to a theoretical threshold of discrimination in the receptor-noise limited</span></span>
<span id="cb233-14"><a href="analysing-data.html#cb233-14" aria-hidden="true" tabindex="-1"></a><span class="co"># model.</span></span>
<span id="cb233-15"><a href="analysing-data.html#cb233-15" aria-hidden="true" tabindex="-1"></a>jnd.mat[<span class="fu">which</span>(jnd.mat <span class="sc">&lt;</span> <span class="dv">1</span>)] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb233-16"><a href="analysing-data.html#cb233-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb233-17"><a href="analysing-data.html#cb233-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Load up a library and use k-means clustering to estimate the number of discrete</span></span>
<span id="cb233-18"><a href="analysing-data.html#cb233-18" aria-hidden="true" tabindex="-1"></a><span class="co"># colours present in our sample.</span></span>
<span id="cb233-19"><a href="analysing-data.html#cb233-19" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(NbClust)</span>
<span id="cb233-20"><a href="analysing-data.html#cb233-20" aria-hidden="true" tabindex="-1"></a>clust <span class="ot">&lt;-</span> <span class="fu">NbClust</span>(jnd.fakescene, <span class="at">diss =</span> jnd.mat, <span class="at">distance =</span> <span class="cn">NULL</span>, <span class="at">method =</span> <span class="st">&#39;centroid&#39;</span>, <span class="at">index =</span> <span class="st">&#39;hartigan&#39;</span>)</span>
<span id="cb233-21"><a href="analysing-data.html#cb233-21" aria-hidden="true" tabindex="-1"></a>clust<span class="sc">$</span>Best.nc</span></code></pre></div>
<pre><code>## Number_clusters     Value_Index 
##           4.000        1639.353</code></pre>
<p>This approach has identified four distinct colour patches, as simulated. To be able to feed this information directly in to <code>adjacent()</code> we need to arrange the now-classified spectral data into a xyz data-image matrix, in which <code>x</code> and <code>y</code> coordinates specify the location of the sample in our scene, and the z values at each location specify the colour-class, or group, into which each sample has been classified (an integer). Many clustering packages offers the data’s classification as output in some form, so we can simply rearrange it into our original 10 x 10 sampling matrix before passing it to <code>adjacent()</code>.</p>
<div class="sourceCode" id="cb235"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb235-1"><a href="analysing-data.html#cb235-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Rearrange the data into a colour-classified image matrix, and take a look at it.</span></span>
<span id="cb235-2"><a href="analysing-data.html#cb235-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Note that is the same structure as the output of &#39;classify()&#39;.</span></span>
<span id="cb235-3"><a href="analysing-data.html#cb235-3" aria-hidden="true" tabindex="-1"></a>mat.fakescene <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">as.numeric</span>(<span class="fu">unlist</span>(clust<span class="sc">$</span>Best.partition)), <span class="dv">10</span>, <span class="dv">10</span>)</span>
<span id="cb235-4"><a href="analysing-data.html#cb235-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(mat.fakescene)</span></code></pre></div>
<pre><code>##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
## [1,]    1    1    1    2    2    3    3    3    4     4
## [2,]    1    1    1    2    2    3    3    3    4     4
## [3,]    1    1    1    2    2    3    3    3    4     4
## [4,]    1    1    1    2    2    3    3    3    4     4
## [5,]    1    1    1    2    2    3    3    3    4     4
## [6,]    1    1    2    2    2    3    3    4    4     4</code></pre>
<div class="sourceCode" id="cb237"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb237-1"><a href="analysing-data.html#cb237-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Run the adjacency analysis</span></span>
<span id="cb237-2"><a href="analysing-data.html#cb237-2" aria-hidden="true" tabindex="-1"></a><span class="fu">adjacent</span>(mat.fakescene, <span class="at">xscale =</span> <span class="dv">200</span>)</span></code></pre></div>
<pre><code>##     k   N n_off  p_1  p_2  p_3  p_4     q_1_1      q_1_2     q_2_2      q_2_3
## img 4 180    32 0.25 0.25 0.25 0.25 0.2055556 0.06111111 0.2055556 0.05555556
##         q_3_3      q_3_4     q_4_4   t_1_2  t_2_3   t_3_4     m   m_r   m_c  A
## img 0.2055556 0.06111111 0.2055556 0.34375 0.3125 0.34375 0.008 0.015 0.001 15
##     Sc       St Jc        Jt  B Rt Rab m_dS s_dS cv_dS m_dL s_dL cv_dL m_hue
## img  4 2.994152  1 0.4990253 NA NA  NA   NA   NA    NA   NA   NA    NA    NA
##     s_hue var_hue m_sat s_sat cv_sat m_lum s_lum cv_lum
## img    NA      NA    NA    NA     NA    NA    NA     NA</code></pre>
<p>The output reflects several aspects of our simulated dataset, which is reassuring. There are four colour classes (<code>k</code>= 4) or equal proportion (<code>p_x</code>), 180 transitions in total (<code>N</code>, 9 transitions per row/column x 10 rows/columns), and the colour diversity (<code>Sc</code>) equals <code>k</code>, and relative colour diversity (<code>Jc</code>) equals 1, since all colours occur in equal proportions.</p>

</div>
</div>
</div>
</div>
<h3>References</h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Andersson2006" class="csl-entry">
Andersson, Staffan, and Maria Prager. 2006. <span>“Quantification of <span>Coloration</span>.”</span> In <em>Bird <span>Coloration</span>, <span>Volume</span> 1: <span>Mechanisms</span> and <span>Measurements</span></em>, edited by Geoffrey E. Hill and Kevin J. McGraw, 1:640. Bird Coloration. Harvard University Press.
</div>
<div id="ref-Backhaus1991" class="csl-entry">
Backhaus, Werner. 1991. <span>“Color Opponent Coding in the Visual System of the Honeybee.”</span> <em>Vision Research</em> 31 (7): 1381–97. <a href="https://doi.org/10.1016/0042-6989(91)90059-E">https://doi.org/10.1016/0042-6989(91)90059-E</a>.
</div>
<div id="ref-Briscoe2001" class="csl-entry">
Briscoe, Adriana D., and Lars Chittka. 2001. <span>“The Evolution of Color Vision in Insects.”</span> <em>Annual Review of Entomology</em> 46 (1): 471–510. <a href="https://doi.org/10.1146/annurev.ento.46.1.471">https://doi.org/10.1146/annurev.ento.46.1.471</a>.
</div>
<div id="ref-Chittka1992" class="csl-entry">
Chittka, Lars. 1992. <span>“The Colour Hexagon: A Chromaticity Diagram Based on Photoreceptor Excitations as a Generalized Representation of Colour Opponency.”</span> <em>Journal of Comparative Physiology A</em> 170 (5): 533–43. <a href="https://doi.org/10.1007/BF00199331">https://doi.org/10.1007/BF00199331</a>.
</div>
<div id="ref-Endler1990" class="csl-entry">
Endler, John A. 1990. <span>“On the Measurement and Classification of Colour in Studies of Animal Colour Patterns.”</span> <em>Biological Journal of the Linnean Society</em> 41 (4): 315–52. <a href="https://doi.org/10.1111/j.1095-8312.1990.tb00839.x">https://doi.org/10.1111/j.1095-8312.1990.tb00839.x</a>.
</div>
<div id="ref-Endler2012" class="csl-entry">
———. 2012. <span>“A Framework for Analysing Colour Pattern Geometry: Adjacent Colours.”</span> <em>Biological Journal of the Linnean Society</em> 107 (2): 233–53. <a href="https://doi.org/10.1111/j.1095-8312.2012.01937.x">https://doi.org/10.1111/j.1095-8312.2012.01937.x</a>.
</div>
<div id="ref-Endler2018" class="csl-entry">
Endler, John A., Gemma L. Cole, and Alexandrea M. Kranz. 2018. <span>“Boundary Strength Analysis: <span>Combining</span> Colour Pattern Geometry and Coloured Patch Visual Properties for Use in Predicting Behaviour and Fitness.”</span> Edited by Simon Blomberg. <em>Methods in Ecology and Evolution</em>, August. <a href="https://doi.org/10.1111/2041-210X.13073">https://doi.org/10.1111/2041-210X.13073</a>.
</div>
<div id="ref-Endler2005" class="csl-entry">
Endler, John A., and Paul W. Mielke. 2005. <span>“Comparing Entire Colour Patterns as Birds See Them.”</span> <em>Biological Journal of the Linnean Society</em> 86 (4): 405–31. <a href="https://doi.org/10.1111/j.1095-8312.2005.00540.x">https://doi.org/10.1111/j.1095-8312.2005.00540.x</a>.
</div>
<div id="ref-Goldsmith1990" class="csl-entry">
Goldsmith, Timothy H. 1990. <span>“Optimization, <span>Constraint</span>, and <span>History</span> in the <span>Evolution</span> of <span>Eyes</span>.”</span> <em>The Quarterly Review of Biology</em> 65 (3): 281–322. <a href="https://doi.org/10.1086/416840">https://doi.org/10.1086/416840</a>.
</div>
<div id="ref-Govardovskii2000" class="csl-entry">
Govardovskii, Victor I., Nanna Fyhrquist, Tom Reuter, Dmitry G. Kuzmin, and Kristian Donner. 2000. <span>“In Search of the Visual Pigment Template.”</span> <em>Visual Neuroscience</em> 17 (4): 509–28. <a href="https://doi.org/10.1017/S0952523800174036">https://doi.org/10.1017/S0952523800174036</a>.
</div>
<div id="ref-Hart2005" class="csl-entry">
Hart, Nathan S., and Misha Vorobyev. 2005. <span>“Modelling Oil Droplet Absorption Spectra and Spectral Sensitivities of Bird Cone Photoreceptors.”</span> <em>Journal of Comparative Physiology A</em> 191 (4): 381–92. <a href="https://doi.org/10.1007/s00359-004-0595-3">https://doi.org/10.1007/s00359-004-0595-3</a>.
</div>
<div id="ref-Maia2018" class="csl-entry">
Maia, Rafael, and Thomas E White. 2018. <span>“Comparing Colors Using Visual Models.”</span> <em>Behavioral Ecology</em> 29 (3): 649–59. <a href="https://doi.org/10.1093/beheco/ary017">https://doi.org/10.1093/beheco/ary017</a>.
</div>
<div id="ref-Montgomerie2006" class="csl-entry">
Montgomerie, Robert. 2006. <span>“Analyzing <span>Colors</span>.”</span> In <em>Bird <span>Coloration</span>, <span>Volume</span> 1: <span>Mechanisms</span> and <span>Measurements</span></em>, edited by Geoffrey E. Hill and Kevin J. McGraw, 1:640. Bird Coloration. Harvard University Press.
</div>
<div id="ref-Renoult2017" class="csl-entry">
Renoult, Julien P., Almut Kelber, and H. Martin Schaefer. 2017. <span>“Colour Spaces in Ecology and Evolutionary Biology.”</span> <em>Biological Reviews</em> 92 (1): 292–315. <a href="https://doi.org/10.1111/brv.12230">https://doi.org/10.1111/brv.12230</a>.
</div>
<div id="ref-Stoddard2008" class="csl-entry">
Stoddard, Mary Caswell, and Richard O. Prum. 2008. <span>“Evolution of Avian Plumage Color in a Tetrahedral Color Space: A Phylogenetic Analysis of <span>New World</span> Buntings.”</span> <em>The American Naturalist</em> 171 (6): 755–76. <a href="https://doi.org/10.1086/587526">https://doi.org/10.1086/587526</a>.
</div>
<div id="ref-Stoddard2011" class="csl-entry">
Stoddard, Mary Caswell, and Martin Stevens. 2011. <span>“Avian Vision and the Evolution of Egg Color Mimicry in the Common Cuckoo.”</span> <em>Evolution</em> 65 (7): 2004–13. <a href="https://doi.org/10.1111/j.1558-5646.2011.01262.x">https://doi.org/10.1111/j.1558-5646.2011.01262.x</a>.
</div>
<div id="ref-Troje1993" class="csl-entry">
Troje, Nikolaus. 1993. <span>“Spectral <span>Categories</span> in the <span>Learning Behaviour</span> of <span>Blowflies</span>.”</span> <em>Zeitschrift für Naturforschung C</em> 48 (1-2): 96–104. <a href="https://doi.org/10.1515/znc-1993-1-218">https://doi.org/10.1515/znc-1993-1-218</a>.
</div>
<div id="ref-Vorobyev2001" class="csl-entry">
Vorobyev, Misha, Robert Brandt, Dagmar Peitsch, Simon B Laughlin, and Randolf Menzel. 2001. <span>“Colour Thresholds and Receptor Noise: Behaviour and Physiology Compared.”</span> <em>Vision Research</em> 41 (5): 639–53. <a href="https://doi.org/10.1016/S0042-6989(00)00288-1">https://doi.org/10.1016/S0042-6989(00)00288-1</a>.
</div>
<div id="ref-Vorobyev1998" class="csl-entry">
Vorobyev, Misha, Daniel Colaco Osorio, Andrew T. D. Bennett, N. Justin Marshall, and Innes C. Cuthill. 1998. <span>“Tetrachromacy, Oil Droplets and Bird Plumage Colours.”</span> <em>Journal of Comparative Physiology A</em> 183 (5): 621–33. <a href="https://doi.org/10.1007/s003590050286">https://doi.org/10.1007/s003590050286</a>.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="importing-processing-and-visualising-data.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="spectral-shape-descriptors.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/bisaloo/colsci-book/edit/main/02-analysing.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
